<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/C.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/C.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":true,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="python, shell, R, perl, java">
<meta property="og:type" content="website">
<meta property="og:title" content="Chaopower&#39;s Blog">
<meta property="og:url" content="http://example.com/index.html">
<meta property="og:site_name" content="Chaopower&#39;s Blog">
<meta property="og:description" content="python, shell, R, perl, java">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="Chao Zhang">
<meta property="article:tag" content="生物信息学">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'zh-CN'
  };
</script>

  <title>Chaopower's Blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Chaopower's Blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">chaopower@126.com</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>关于</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>标签<span class="badge">3</span></a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>分类<span class="badge">6</span></a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>归档<span class="badge">5</span></a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2023/03/02/sam-bam%E6%96%87%E4%BB%B6/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Chao Zhang">
      <meta itemprop="description" content="python, shell, R, perl, java">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Chaopower's Blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2023/03/02/sam-bam%E6%96%87%E4%BB%B6/" class="post-title-link" itemprop="url">sam-bam文件</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2023-03-02 21:19:08 / 修改时间：21:37:06" itemprop="dateCreated datePublished" datetime="2023-03-02T21:19:08+08:00">2023-03-02</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6/" itemprop="url" rel="index"><span itemprop="name">生信软件</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>SAM(Sequence Alignment&#x2F;Map)格式是一种通用的比对格式，用来存储reads与参考序列的比对信息。SAM是一种序列比对格式标准，由sanger制定，是以TAB为分割符的文本格式。主要应用于测序序列mapping到基因组上的结果表示。 SAM分为两部分，注释信息（header section）和比对结果部分（alignment section）。</p>
<h5 id="1-注释信息"><a href="#1-注释信息" class="headerlink" title="1.注释信息"></a>1.注释信息</h5><p>header内容不多，也不会太复杂，每一行都用‘@’ 符号开头，里面主要包含了版本信息，序列比对的参考序列信息，如果是标准工具（bwa，bowtie，picard）生成的SAM，一般还会包含生成该份文件的参数信息（如下图），@PG标签开头的那些。</p>
<p><img src="/sam-bam%E6%96%87%E4%BB%B6/image-20230302212240356.png" alt="image-20230302212240356"></p>
<h5 id="2-比对信息"><a href="#2-比对信息" class="headerlink" title="2. 比对信息"></a>2. 比对信息</h5><p>这是我们通常所说的序列比对内容，每一行都是一条read比对信息，所有列都以TAB分隔，它的记录看起来是这样的：</p>
<p><img src="/sam-bam%E6%96%87%E4%BB%B6/image-20230302213359578.png" alt="image-20230302213359578"></p>
<p>下面我们就来仔细瞧瞧这里的每一个信息分别都是什么。</p>
<p>以下是相关信息的具体说明：</p>
<ul>
<li><p>flag信息</p>
<p>这是一个非常特别并且重要的数字，也是一个容易被忽视的数字，它里面实际上记录了许多有关read比对情况的信息。想要读懂它的一个关键点是我们不能够将其视为一个数字，而是必须将其转换为一串由0和1组成的二进制码，这一串二进制数中的每一个位(注意是“位”，bit的意思)都代表了一个特定信息，它一共有12位（以前只有8位），所以一般会用一个16位的整数来代表，这个整数的值就是12个0和1的组合计算得来的，因此它的数值范围是0~2048（2的12次方），如下表：</p>
</li>
</ul>
<p><img src="/sam-bam%E6%96%87%E4%BB%B6/image-20230302213545662.png" alt="image-20230302213545662"></p>
<p>以下是相关信息的具体说明：</p>
<ul>
<li><p>flag信息</p>
<p>这是一个非常特别并且重要的数字，也是一个容易被忽视的数字，它里面实际上记录了许多有关read比对情况的信息。想要读懂它的一个关键点是我们不能够将其视为一个数字，而是必须将其转换为一串由0和1组成的二进制码，这一串二进制数中的每一个位(注意是“位”，bit的意思)都代表了一个特定信息，它一共有12位（以前只有8位），所以一般会用一个16位的整数来代表，这个整数的值就是12个0和1的组合计算得来的，因此它的数值范围是0~2048（2的12次方），如下表：</p>
<p><img src="/sam-bam%E6%96%87%E4%BB%B6/image-20230302213625278.png" alt="image-20230302213625278"></p>
<p>所以，通过上面这个表的信息，我们就可以清楚地知道每一个FLAG中都包含了什么信息。比如看到FLAG &#x3D; 77时，我们第一步要做的就是将其分解为二进制序列（也可以理解为分解成若干个2的n次方之和）：</p>
<p>77 &#x3D; 000001001101 &#x3D; 1 + 4 + 8 +64，这样就得到了这个FLAG包含的意思：PE read，read比对不上参考序列，它的配对read也同样比不上参考序列，它是read1。</p>
<p>此处提供一个在线换算flag值的网站：<a target="_blank" rel="noopener" href="https://broadinstitute.github.io/picard/explain-flags.html">https://broadinstitute.github.io/picard/explain-flags.html</a></p>
</li>
</ul>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2023/03/01/samtools%20%E4%BD%BF%E7%94%A8/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Chao Zhang">
      <meta itemprop="description" content="python, shell, R, perl, java">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Chaopower's Blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2023/03/01/samtools%20%E4%BD%BF%E7%94%A8/" class="post-title-link" itemprop="url">生信软件-samtools</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2023-03-01 21:02:58 / 修改时间：19:10:14" itemprop="dateCreated datePublished" datetime="2023-03-01T21:02:58+08:00">2023-03-01</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6/" itemprop="url" rel="index"><span itemprop="name">生信软件</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h1 id="samtools-常用命令"><a href="#samtools-常用命令" class="headerlink" title="samtools 常用命令"></a>samtools 常用命令</h1><h2 id="1-软件介绍及安装"><a href="#1-软件介绍及安装" class="headerlink" title="1.软件介绍及安装"></a>1.软件介绍及安装</h2><p>Samtools 用于处理sam与bam格式的工具软件，能够实现二进制查看、格式转换排序及合并等功能，结合sam格式中的flag、tag等信息，还可以完成比对结果的统计汇总，同时利用linux中的grep、awk等操作命令，还可以大大扩展samtools的使用范围与功能。</p>
<figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#获取压缩包并解压</span></span><br><span class="line">wget</span><br><span class="line">https://github.com/samtools/samtools/releases/download/1.9/samtools</span><br><span class="line">1.9.tar.bz2</span><br><span class="line">tar -xzvf samtools-1.9.tar .bz2</span><br><span class="line"><span class="comment">#进入samtools文件夹</span></span><br><span class="line"><span class="built_in">cd</span> samtools-1.9</span><br><span class="line"><span class="comment">#安装</span></span><br><span class="line">./configure</span><br><span class="line">make</span><br><span class="line">sudo make install</span><br><span class="line"><span class="comment">##非root安装</span></span><br><span class="line"><span class="comment">#./configure --prefix=/home/用户名/.Local</span></span><br><span class="line"><span class="comment">#make</span></span><br><span class="line"><span class="comment">#make install</span></span><br></pre></td></tr></table></figure>

<h2 id="2-Samtools-基本命令"><a href="#2-Samtools-基本命令" class="headerlink" title="2.Samtools 基本命令"></a>2.Samtools 基本命令</h2><figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line">$ samtools --<span class="built_in">help</span></span><br><span class="line"></span><br><span class="line">Program: samtools (Tools <span class="keyword">for</span> alignments <span class="keyword">in</span> the SAM format)</span><br><span class="line">Version: 1.3.1 (using htslib 1.3.1)</span><br><span class="line"></span><br><span class="line">Usage:   samtools &lt;<span class="built_in">command</span>&gt; [options]</span><br><span class="line"></span><br><span class="line">Commands:</span><br><span class="line">  -- Indexing</span><br><span class="line">     dict           create a sequence dictionary file</span><br><span class="line">     faidx          index/extract FASTA</span><br><span class="line">     index          index alignment</span><br><span class="line"></span><br><span class="line">  -- Editing</span><br><span class="line">     calmd          recalculate MD/NM tags and <span class="string">&#x27;=&#x27;</span> bases</span><br><span class="line">     fixmate        fix mate information</span><br><span class="line">     reheader       replace BAM header</span><br><span class="line">     rmdup          remove PCR duplicates</span><br><span class="line">     targetcut      <span class="built_in">cut</span> fosmid regions (<span class="keyword">for</span> fosmid pool only)</span><br><span class="line">     addreplacerg   adds or replaces RG tags</span><br><span class="line"></span><br><span class="line">  -- File operations</span><br><span class="line">     collate        shuffle and group alignments by name</span><br><span class="line">     <span class="built_in">cat</span>            concatenate BAMs</span><br><span class="line">     merge          merge sorted alignments</span><br><span class="line">     mpileup        multi-way pileup</span><br><span class="line">     <span class="built_in">sort</span>           <span class="built_in">sort</span> alignment file</span><br><span class="line">     <span class="built_in">split</span>          splits a file by <span class="built_in">read</span> group</span><br><span class="line">     quickcheck     quickly check <span class="keyword">if</span> SAM/BAM/CRAM file appears intact</span><br><span class="line">     fastq          converts a BAM to a FASTQ</span><br><span class="line">     fasta          converts a BAM to a FASTA</span><br><span class="line"></span><br><span class="line">  -- Statistics</span><br><span class="line">     bedcov         <span class="built_in">read</span> depth per BED region</span><br><span class="line">     depth          compute the depth</span><br><span class="line">     flagstat       simple stats</span><br><span class="line">     idxstats       BAM index stats</span><br><span class="line">     phase          phase heterozygotes</span><br><span class="line">     stats          generate stats (former bamcheck)</span><br><span class="line"></span><br><span class="line">  -- Viewing</span><br><span class="line">     flags          explain BAM flags</span><br><span class="line">     tview          text alignment viewer</span><br><span class="line">     view           SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion</span><br><span class="line">     depad          convert padded BAM to unpadded BAM</span><br></pre></td></tr></table></figure>

<h3 id="2-1-构建索引"><a href="#2-1-构建索引" class="headerlink" title="2.1 构建索引"></a>2.1 构建索引</h3><h4 id="2-1-1-dict"><a href="#2-1-1-dict" class="headerlink" title="2.1.1 dict"></a>2.1.1 dict</h4><p><strong>dict</strong> – 对fasta格式的参考基因组序列构建字典索引</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ samtools dict</span><br><span class="line"></span><br><span class="line">About:   Create a sequence dictionary file from a fasta file</span><br><span class="line">Usage:   samtools dict [options] &lt;file.fa|file.fa.gz&gt;</span><br><span class="line"></span><br><span class="line">Options: -a, --assembly STR    assembly</span><br><span class="line">         -H, --no-header       do not print @HD line</span><br><span class="line">         -o, --output STR      file to write out dict file [stdout]</span><br><span class="line">         -s, --species STR     species</span><br><span class="line">         -u, --uri STR         URI [file:///abs/path/to/file.fa]</span><br></pre></td></tr></table></figure>

<p>Example：<code>samtools dict -a GRCh38 -s &quot;Homo sapiens&quot; hg38.fasta -o hg38.dict</code></p>
<p>构建完成后会生成一个hg38.dict的参考基因组索引文件</p>
<h4 id="2-1-2-faidx"><a href="#2-1-2-faidx" class="headerlink" title="2.1.2  faidx"></a>2.1.2  faidx</h4><p><strong>faidx</strong> – 对fasta格式的文件建立索引，生成的索引文件以.fai后缀结尾。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools faidx</span><br><span class="line"></span><br><span class="line">Usage:   samtools faidx &lt;file.fa|file.fa.gz&gt; [&lt;reg&gt; [...]]</span><br></pre></td></tr></table></figure>

<p>Example：<code>samtools faidx hg38.fasta</code></p>
<p>构建完成后会生成一个hg38.fasta.fai的索引文件，该文件共有五列，分别表示：</p>
<p>第一列：序列的名称</p>
<p>第二列：序列的长度</p>
<p>第三列：第一个碱基的偏移量，从0开始计数</p>
<p>第四列：除了最后一行外，序列中每行的碱基数</p>
<p>第五列：除了最后一行外，序列中每行的长度（包括换行符）</p>
<p><strong>该命令也可以根据索引文件，快速提取fasta文件中任意区域的序列文件</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 提取1号染色体的序列信息</span><br><span class="line">$ samtools faidx ha38.fasta chr1 &gt; hg38.chr1.fa</span><br><span class="line"></span><br><span class="line"># 提取1号染色体200-1000bp之间的序列</span><br><span class="line">$ samtools faidx hg38.fasta chr1:200-1000 &gt; hg38.chr1_200_1000.fa</span><br></pre></td></tr></table></figure>

<h4 id="2-1-3-index"><a href="#2-1-3-index" class="headerlink" title="2.1.3 index"></a>2.1.3 index</h4><p> index – 对bam格式的比对文件构建索引，需对bam文件进行排序后才能构建索引</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ samtools index</span><br><span class="line"></span><br><span class="line">Usage: samtools index [-bc] [-m INT] &lt;in.bam&gt; [out.index]</span><br><span class="line">Options:</span><br><span class="line">  -b       Generate BAI-format index for BAM files [default]</span><br><span class="line">  -c       Generate CSI-format index for BAM files</span><br><span class="line">  -m INT   Set minimum interval size for CSI indices to 2^INT [14]</span><br></pre></td></tr></table></figure>

<p>Example： <code>samtools index aln.sorted.bam</code></p>
<p>构建完成后将产生后缀为.bai的文件，用于快速的随机处理</p>
<h3 id="2-2-文件编辑-Editing"><a href="#2-2-文件编辑-Editing" class="headerlink" title="2.2 文件编辑(Editing)"></a>2.2 文件编辑(Editing)</h3><h4 id="calmd"><a href="#calmd" class="headerlink" title="calmd"></a>calmd</h4><ul>
<li>– recalculate MD&#x2F;NM tags and ‘&#x3D;’ bases 计算MD tag标签值（ 记录了mismatch信息 ）</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools calmd</span><br><span class="line">Usage: samtools calmd [-eubrAES] &lt;aln.bam&gt; &lt;ref.fasta&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -e       change identical bases to <span class="string">&#x27;=&#x27;</span></span><br><span class="line">  -u       uncompressed BAM output (<span class="keyword">for</span> piping)</span><br><span class="line">  -b       compressed BAM output</span><br><span class="line">  -S       ignored (input format is auto-detected)</span><br><span class="line">  -A       modify the quality string</span><br><span class="line">  -r       compute the BQ tag (without -A) or <span class="built_in">cap</span> baseQ by BAQ (with -A)</span><br><span class="line">  -E       extended BAQ <span class="keyword">for</span> better sensitivity but lower specificity</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example：<code>samtools calmd -AEur aln.bam ref.fa</code></li>
</ul>
<h4 id="fixmate"><a href="#fixmate" class="headerlink" title="fixmate"></a>fixmate</h4><ul>
<li>– fix mate information  为以名称排序定位的alignment填入配对坐标，ISIZE（inferred insert size推测的插入序列大小）和配对相应的标签（flag）</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools fixmate</span><br><span class="line"></span><br><span class="line">Usage: samtools fixmate &lt;in.nameSrt.bam&gt; &lt;out.nameSrt.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -r           Remove unmapped reads and secondary alignments</span><br><span class="line">  -p           Disable FR proper pair check</span><br><span class="line">  -c           Add template cigar ct tag</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">As elsewhere <span class="keyword">in</span> samtools, use <span class="string">&#x27;-&#x27;</span> as the filename <span class="keyword">for</span> stdin/stdout. The input</span><br><span class="line">file must be grouped by <span class="built_in">read</span> name (e.g. sorted by name). Coordinated sorted</span><br><span class="line">input is not accepted.</span><br></pre></td></tr></table></figure>

<ul>
<li>Example：<code>samtools fixmate in.namesorted.sam out.bam</code></li>
</ul>
<h4 id="reheader"><a href="#reheader" class="headerlink" title="reheader"></a>reheader</h4><p> – replace BAM header 替换bam文件头注释信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ samtools reheader</span><br><span class="line"></span><br><span class="line">Usage: samtools reheader [-P] in.header.sam in.bam &gt; out.bam</span><br><span class="line">   or  samtools reheader [-P] -i in.header.sam file.bam</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">    -P, --no-PG      Do not generate an @PG header line.</span><br><span class="line">    -i, --in-place   Modify the bam/cram file directly.</span><br><span class="line">                     (Defaults to outputting to stdout.)</span><br></pre></td></tr></table></figure>

<h4 id="rmdup"><a href="#rmdup" class="headerlink" title="rmdup"></a>rmdup</h4><ul>
<li>– remove PCR duplicates 去除bam文件中的PCR重复信息</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">$ samtools rmdup</span><br><span class="line"></span><br><span class="line">Usage:  samtools rmdup [-sS] &lt;input.srt.bam&gt; &lt;output.bam&gt;</span><br><span class="line"></span><br><span class="line">Option: -s    rmdup <span class="keyword">for</span> SE reads</span><br><span class="line">        -S    treat PE reads as SE <span class="keyword">in</span> rmdup (force -s)</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example：<code>samtools rmdup -sS in.sorted.bam output.bam</code></li>
</ul>
<h4 id="targetcut"><a href="#targetcut" class="headerlink" title="targetcut"></a>targetcut</h4><ul>
<li><p>– cut fosmid regions (for fosmid pool only)   此命令仅用于从fosmid混池测序中切除fosmid克隆序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">$ samtools targetcut</span><br><span class="line"></span><br><span class="line">Usage: samtools targetcut [-Q minQ] [-i inPen] [-0 em0] [-1 em1] [-2 em2] &lt;in.bam&gt;</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -f, --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure></li>
</ul>
<h4 id="addreplacerg"><a href="#addreplacerg" class="headerlink" title="addreplacerg"></a>addreplacerg</h4><p>– adds or replaces RG tags 该命令用于添加或更改bam文件中的RG tag标签值</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools addreplacerg</span><br><span class="line"></span><br><span class="line">Usage: samtools addreplacerg [options] [-r &lt;@RG line&gt; | -R &lt;existing <span class="built_in">id</span>&gt;] [-o &lt;output.bam&gt;] &lt;input.bam&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -m MODE   Set the mode of operation from one of overwrite_all, orphan_only [overwrite_all]</span><br><span class="line">  -o FILE   Where to write output to [stdout]</span><br><span class="line">  -r STRING @RG line text</span><br><span class="line">  -R STRING ID of @RG line <span class="keyword">in</span> existing header to use</span><br><span class="line">      --input-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify input format (SAM, BAM, CRAM)</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example ：<code>samtools addreplacerg -r &#39;ID:fish&#39; -r &#39;LB:1334&#39; -r &#39;SM:alpha&#39; -o output.bam input.bam</code></li>
</ul>
<h3 id="2-3-文件处理-File-operations"><a href="#2-3-文件处理-File-operations" class="headerlink" title="2.3 文件处理(File operations)"></a>2.3 文件处理(File operations)</h3><h4 id="collate"><a href="#collate" class="headerlink" title="collate"></a>collate</h4><p>– shuffle and group alignments by name 根据read name信息将bam文件进行随机打散和分组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">$ samtools collate</span><br><span class="line"></span><br><span class="line">Usage:   samtools collate [-Ou] [-n nFiles] [-c cLevel] &lt;in.bam&gt; &lt;out.prefix&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">      -O       output to stdout</span><br><span class="line">      -u       uncompressed BAM output</span><br><span class="line">      -l INT   compression level [1]</span><br><span class="line">      -n INT   number of temporary files [64]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools collate -o aln.name_collated.bam aln.sorted.bam</code></li>
</ul>
<h4 id="cat"><a href="#cat" class="headerlink" title="cat"></a>cat</h4><p> – concatenate BAMs  将多个bam文件合并为一个bam文件（与merge命令的区别是cat不需要将bam文件提前进行排序）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools <span class="built_in">cat</span></span><br><span class="line"></span><br><span class="line">Usage: samtools <span class="built_in">cat</span> [-h header.sam] [-o out.bam] &lt;in1.bam&gt; [...]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools cat -o out.bam in1.bam in2.bam</code></li>
</ul>
<h4 id="merge"><a href="#merge" class="headerlink" title="merge"></a>merge</h4><ul>
<li>– merge sorted alignments 将多个已经sort的bam文件进行合并</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">$ samtools merge</span><br><span class="line"></span><br><span class="line">Usage: samtools merge [-nurlf] [-h inh.sam] [-b &lt;bamlist.fofn&gt;] &lt;out.bam&gt; &lt;in1.bam&gt; [&lt;in2.bam&gt; ... &lt;inN.bam&gt;]</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -n         Input files are sorted by <span class="built_in">read</span> name</span><br><span class="line">  -r         Attach RG tag (inferred from file names)</span><br><span class="line">  -u         Uncompressed BAM output</span><br><span class="line">  -f         Overwrite the output BAM <span class="keyword">if</span> exist</span><br><span class="line">  -1         Compress level 1</span><br><span class="line">  -l INT     Compression level, from 0 to 9 [-1]</span><br><span class="line">  -R STR     Merge file <span class="keyword">in</span> the specified region STR [all]</span><br><span class="line">  -h FILE    Copy the header <span class="keyword">in</span> FILE to &lt;out.bam&gt; [in1.bam]</span><br><span class="line">  -c         Combine @RG headers with colliding IDs [alter IDs to be distinct]</span><br><span class="line">  -p         Combine @PG headers with colliding IDs [alter IDs to be distinct]</span><br><span class="line">  -s VALUE   Override random seed</span><br><span class="line">  -b FILE    List of input BAM filenames, one per line [null]</span><br><span class="line">  -@, --threads INT</span><br><span class="line">             Number of BAM/CRAM compression threads [0]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools merge merged.bam in1.sorted.bam in2.sorted.bam</code></li>
</ul>
<h4 id="mpileup"><a href="#mpileup" class="headerlink" title="mpileup"></a>mpileup</h4><p> – multi-way pileup  对参考基因组每个位点做碱基堆积，用于call SNP和INDEL。主要是生成BCF、VCF文件或者pileup一个或多个bam文件。比对记录以在@RG中的样本名作为区分标识符。如果样本标识符缺失，那么每一个输入文件则视为一个样本</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br></pre></td><td class="code"><pre><span class="line">$ samtools mpileup</span><br><span class="line"></span><br><span class="line">Usage: samtools mpileup [options] in1.bam [in2.bam [...]]</span><br><span class="line"></span><br><span class="line">Input options:</span><br><span class="line">  -6, --illumina1.3+      quality is <span class="keyword">in</span> the Illumina-1.3+ encoding</span><br><span class="line">  -A, --count-orphans     <span class="keyword">do</span> not discard anomalous <span class="built_in">read</span> pairs</span><br><span class="line">  -b, --bam-list FILE     list of input BAM filenames, one per line</span><br><span class="line">  -B, --no-BAQ            <span class="built_in">disable</span> BAQ (per-Base Alignment Quality)</span><br><span class="line">  -C, --adjust-MQ INT     adjust mapping quality; recommended:50, <span class="built_in">disable</span>:0 [0]</span><br><span class="line">  -d, --max-depth INT     max per-file depth; avoids excessive memory usage [250]</span><br><span class="line">  -E, --redo-BAQ          recalculate BAQ on the fly, ignore existing BQs</span><br><span class="line">  -f, --fasta-ref FILE    faidx indexed reference sequence file</span><br><span class="line">  -G, --exclude-RG FILE   exclude <span class="built_in">read</span> <span class="built_in">groups</span> listed <span class="keyword">in</span> FILE</span><br><span class="line">  -l, --positions FILE    skip unlisted positions (chr pos) or regions (BED)</span><br><span class="line">  -q, --min-MQ INT        skip alignments with mapQ smaller than INT [0]</span><br><span class="line">  -Q, --min-BQ INT        skip bases with baseQ/BAQ smaller than INT [13]</span><br><span class="line">  -r, --region REG        region <span class="keyword">in</span> <span class="built_in">which</span> pileup is generated</span><br><span class="line">  -R, --ignore-RG         ignore RG tags (one BAM = one sample)</span><br><span class="line">  --rf, --incl-flags STR|INT  required flags: skip reads with mask bits <span class="built_in">unset</span> []</span><br><span class="line">  --ff, --excl-flags STR|INT  filter flags: skip reads with mask bits <span class="built_in">set</span></span><br><span class="line">                                            [UNMAP,SECONDARY,QCFAIL,DUP]</span><br><span class="line">  -x, --ignore-overlaps   <span class="built_in">disable</span> read-pair overlap detection</span><br><span class="line"></span><br><span class="line">Output options:</span><br><span class="line">  -o, --output FILE       write output to FILE [standard output]</span><br><span class="line">  -g, --BCF               generate genotype likelihoods <span class="keyword">in</span> BCF format</span><br><span class="line">  -v, --VCF               generate genotype likelihoods <span class="keyword">in</span> VCF format</span><br><span class="line"></span><br><span class="line">Output options <span class="keyword">for</span> mpileup format (without -g/-v):</span><br><span class="line">  -O, --output-BP         output base positions on reads</span><br><span class="line">  -s, --output-MQ         output mapping quality</span><br><span class="line"></span><br><span class="line">Output options <span class="keyword">for</span> genotype likelihoods (when -g/-v is used):</span><br><span class="line">  -t, --output-tags LIST  optional tags to output:</span><br><span class="line">               DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []</span><br><span class="line">  -u, --uncompressed      generate uncompressed VCF/BCF output</span><br><span class="line"></span><br><span class="line">SNP/INDEL genotype likelihoods options (effective with -g/-v):</span><br><span class="line">  -e, --ext-prob INT      Phred-scaled gap extension <span class="built_in">seq</span> error probability [20]</span><br><span class="line">  -F, --gap-frac FLOAT    minimum fraction of gapped reads [0.002]</span><br><span class="line">  -h, --tandem-qual INT   coefficient <span class="keyword">for</span> homopolymer errors [100]</span><br><span class="line">  -I, --skip-indels       <span class="keyword">do</span> not perform indel calling</span><br><span class="line">  -L, --max-idepth INT    maximum per-file depth <span class="keyword">for</span> INDEL calling [250]</span><br><span class="line">  -m, --min-ireads INT    minimum number gapped reads <span class="keyword">for</span> indel candidates [1]</span><br><span class="line">  -o, --open-prob INT     Phred-scaled gap open <span class="built_in">seq</span> error probability [40]</span><br><span class="line">  -p, --per-sample-mF     apply -m and -F per-sample <span class="keyword">for</span> increased sensitivity</span><br><span class="line">  -P, --platforms STR     comma separated list of platforms <span class="keyword">for</span> indels [all]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">Notes: Assuming diploid individuals.</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example: <code>samtools mpileup -C 50 -f ref.fasta -r chr3:1,000-2,000 in1.bam in2.bam</code></p>
</li>
<li><p>主要参数解读：</p>
<p>-A：在检测变异中，不忽略异常的reads对</p>
<p>-C：用于调节比对质量的系数，如果reads中含有过多的错配，不能设置为零</p>
<p>-D：输出每个样本的reads深度</p>
<p>-l：BED文件或者包含区域位点的位置列表文件</p>
<p>注意：位置文件包含两列，染色体和位置，从1开始计数。BED文件至少包含3列，染色体、起始和终止位置，开始端从0开始计数。</p>
<p>-r：在指定区域产生pileup，需已建立索引的bam文件，通常和-l参数一起使用</p>
<p>-o&#x2F;g&#x2F;v：输出文件类型（标准格式文件或者VCF、BCF文件）</p>
<p>-t：设置FORMAT和INFO的列表内容，以逗号分割</p>
<p>-u：生成未压缩的VCF和BCF文件</p>
<p>-I：跳过INDEL检测</p>
<p>-m：候选INDEL的最小间隔的reads</p>
<p>-f：输入有索引文件的fasta参考序列</p>
<p>-F ：含有间隔reads的最小片段</p>
</li>
<li><p>mpileup生成的结果包含6列：1) 参考序列名，2) 位置，3) 参考碱基，4) 比对上的reads数，5) 比对情况，6) 比对上的碱基的质量。其中第5列比较复杂,解释如下：<br> a) ‘.’代表与参考序列正链匹配。<br> b) ‘,’代表与参考序列负链匹配。<br> c) ‘ATCGN’代表在正链上的不匹配。<br> d) ‘atcgn’代表在负链上的不匹配。<br> e) ‘*’代表模糊碱基<br> f) ‘’代表匹配的碱基是一个read的开始；’’后面紧跟的ascii码减去33代表比对质量；这两个符号修饰的是后面的碱基，其后紧跟的碱基(.,ATCGatcgNn)代表该read的第一个碱基。<br> g) ‘$’代表一个read的结束，该符号修饰的是其前面的碱基。<br> h) 正则式’+[0-9]+[ACGTNacgtn]+’代表在该位点后插入的碱基；<br> i) 正则式’-[0-9]+[ACGTNacgtn]+’代表在该位点后缺失的碱基；</p>
</li>
</ul>
<h4 id="sort"><a href="#sort" class="headerlink" title="sort"></a>sort</h4><p>– sort alignment file 对比对后的bam文件进行排序，默认按coordinate进行排序</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools <span class="built_in">sort</span></span><br><span class="line"></span><br><span class="line">Usage: samtools <span class="built_in">sort</span> [options...] [in.bam]</span><br><span class="line">Options:</span><br><span class="line">  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)</span><br><span class="line">  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]</span><br><span class="line">  -n         Sort by <span class="built_in">read</span> name</span><br><span class="line">  -o FILE    Write final output to FILE rather than standard output</span><br><span class="line">  -T PREFIX  Write temporary files to PREFIX.nnnn.bam</span><br><span class="line">  -@, --threads INT</span><br><span class="line">             Set number of sorting and compression threads [1]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example: <code>samtools sort -T /tmp/aln.sorted -o aln.sorted.bam aln.bam</code></p>
</li>
<li><p>主要参数释义：</p>
<p>-l：设置文件压缩等级，0不压缩，9压缩最高</p>
<p>-m：每个线程运行内存大小（可使用K M G表示）</p>
<p>-n：按照read名称进行排序</p>
<p>-o：排序后的输出文件</p>
<p>-T：PREFIX临时文件前缀</p>
<p>-@：设置排序和压缩的线程数，默认单线程</p>
</li>
</ul>
<h4 id="split"><a href="#split" class="headerlink" title="split"></a>split</h4><ul>
<li><p>– splits a file by read group 根据read group标签将bam文件进行分割</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">$ samtools <span class="built_in">split</span></span><br><span class="line"></span><br><span class="line">Usage: samtools <span class="built_in">split</span> [-u &lt;unaccounted.bam&gt;[:&lt;unaccounted_header.sam&gt;]]</span><br><span class="line">                      [-f &lt;format_string&gt;] [-v] &lt;merged.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -f STRING       output filename format string [<span class="string">&quot;%*_%#.%.&quot;</span>]</span><br><span class="line">  -u FILE1        put reads with no RG tag or an unrecognised RG tag <span class="keyword">in</span> FILE1</span><br><span class="line">  -u FILE1:FILE2  ...and override the header with FILE2</span><br><span class="line">  -v              verbose output</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">Format string expansions:</span><br><span class="line">  %%     %</span><br><span class="line">  %*     <span class="built_in">basename</span></span><br><span class="line">  %<span class="comment">#     @RG index</span></span><br><span class="line">  %!     @RG ID</span><br><span class="line">  %.     filename extension <span class="keyword">for</span> output format</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools split merged.bam</code></li>
</ul>
</li>
</ul>
<h4 id="quickcheck"><a href="#quickcheck" class="headerlink" title="quickcheck"></a>quickcheck</h4><ul>
<li><p>– quickly check if SAM&#x2F;BAM&#x2F;CRAM file appears intact  检查SAM&#x2F;BAM&#x2F;CRAM文件的完整性</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools quickcheck</span><br><span class="line"></span><br><span class="line">Usage: samtools quickcheck [options] &lt;input&gt; [...]</span><br><span class="line">Options:</span><br><span class="line">  -v              verbose output (repeat <span class="keyword">for</span> more verbosity)</span><br><span class="line"></span><br><span class="line">Notes:</span><br><span class="line"></span><br><span class="line">1. In order to use this <span class="built_in">command</span> effectively, you should check its <span class="built_in">exit</span> status;</span><br><span class="line">   without any -v options it will NOT <span class="built_in">print</span> any output, even when some files</span><br><span class="line">   fail the check. One way to use quickcheck might be as a check that all</span><br><span class="line">   BAM files <span class="keyword">in</span> a directory are okay:</span><br><span class="line"></span><br><span class="line">        samtools quickcheck *.bam &amp;&amp; <span class="built_in">echo</span> <span class="string">&#x27;all ok&#x27;</span> \</span><br><span class="line">           || <span class="built_in">echo</span> <span class="string">&#x27;fail!&#x27;</span></span><br><span class="line"></span><br><span class="line">   To also determine <span class="built_in">which</span> files have failed, use the -v option:</span><br><span class="line"></span><br><span class="line">        samtools quickcheck -v *.bam &gt; bad_bams.fofn \</span><br><span class="line">           &amp;&amp; <span class="built_in">echo</span> <span class="string">&#x27;all ok&#x27;</span> \</span><br><span class="line">           || <span class="built_in">echo</span> <span class="string">&#x27;some files failed check, see bad_bams.fofn&#x27;</span></span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools quickcheck in1.bam in2.cram</code></li>
</ul>
</li>
</ul>
<h4 id="fastq"><a href="#fastq" class="headerlink" title="fastq"></a>fastq</h4><ul>
<li><p>– converts a BAM to a FASTQ 将bam格式文件转换为fastq文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">$ samtools fastq</span><br><span class="line"></span><br><span class="line">Usage: samtools fastq [options...] &lt;in.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -0 FILE   write paired reads flagged both or neither READ1 and READ2 to FILE</span><br><span class="line">  -1 FILE   write paired reads flagged READ1 to FILE</span><br><span class="line">  -2 FILE   write paired reads flagged READ2 to FILE</span><br><span class="line">  -f INT    only include reads with all bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -F INT    only include reads with none of the bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -n        don<span class="string">&#x27;t append /1 and /2 to the read name</span></span><br><span class="line"><span class="string">  -O        output quality in the OQ tag if present</span></span><br><span class="line"><span class="string">  -s FILE   write singleton reads to FILE [assume single-end]</span></span><br><span class="line"><span class="string">  -t        copy RG, BC and QT tags to the FASTQ header line</span></span><br><span class="line"><span class="string">  -v INT    default quality score if not given in file [1]</span></span><br><span class="line"><span class="string">      --input-fmt-option OPT[=VAL]</span></span><br><span class="line"><span class="string">               Specify a single input file format option in the form</span></span><br><span class="line"><span class="string">               of OPTION or OPTION=VALUE</span></span><br><span class="line"><span class="string">      --reference FILE</span></span><br><span class="line"><span class="string">               Reference sequence FASTA FILE [null]</span></span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools fastq input.bam &gt; output.fastq</code></li>
</ul>
</li>
</ul>
<h4 id="fasta"><a href="#fasta" class="headerlink" title="fasta"></a>fasta</h4><p> – converts a BAM to a FASTA 将bam格式文件转换为fasta文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">$ samtools fasta</span><br><span class="line"></span><br><span class="line">Usage: samtools fasta [options...] &lt;in.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -0 FILE   write paired reads flagged both or neither READ1 and READ2 to FILE</span><br><span class="line">  -1 FILE   write paired reads flagged READ1 to FILE</span><br><span class="line">  -2 FILE   write paired reads flagged READ2 to FILE</span><br><span class="line">  -f INT    only include reads with all bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -F INT    only include reads with none of the bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -n        don<span class="string">&#x27;t append /1 and /2 to the read name</span></span><br><span class="line"><span class="string">  -s FILE   write singleton reads to FILE [assume single-end]</span></span><br><span class="line"><span class="string">  -t        copy RG, BC and QT tags to the FASTA header line</span></span><br><span class="line"><span class="string">      --input-fmt-option OPT[=VAL]</span></span><br><span class="line"><span class="string">               Specify a single input file format option in the form</span></span><br><span class="line"><span class="string">               of OPTION or OPTION=VALUE</span></span><br><span class="line"><span class="string">      --reference FILE</span></span><br><span class="line"><span class="string">               Reference sequence FASTA FILE [null]</span></span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools fasta input.bam &gt; output.fasta</code></li>
</ul>
<h3 id="2-4-数据统计-Statistics"><a href="#2-4-数据统计-Statistics" class="headerlink" title="2.4 数据统计(Statistics)"></a>2.4 数据统计(Statistics)</h3><h4 id="bedcov"><a href="#bedcov" class="headerlink" title="bedcov"></a>bedcov</h4><ul>
<li><p>– read depth per BED region 统计给定region区间内reads的深度，每个碱基的深度叠加在一起</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ samtools bedcov</span><br><span class="line"></span><br><span class="line">Usage: samtools bedcov [options] &lt;in.bed&gt; &lt;in1.bam&gt; [...]</span><br><span class="line"></span><br><span class="line">  -Q INT       Only count bases of at least INT quality [0]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools bedcov in.bed aln.sorted.bam</code></li>
</ul>
</li>
</ul>
<h4 id="depth"><a href="#depth" class="headerlink" title="depth"></a>depth</h4><ul>
<li><p>– compute the depth 统计每个碱基位点的测序深度，注意计算前必须对bam文件排序并构建索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">$ samtools depth</span><br><span class="line"></span><br><span class="line">Usage: samtools depth [options] in1.bam [in2.bam [...]]</span><br><span class="line">Options:</span><br><span class="line">   -a                  output all positions (including zero depth)</span><br><span class="line">   -a -a (or -aa)      output absolutely all positions, including unused ref. sequences</span><br><span class="line">   -b &lt;bed&gt;            list of positions or regions</span><br><span class="line">   -f &lt;list&gt;           list of input BAM filenames, one per line [null]</span><br><span class="line">   -l &lt;int&gt;            <span class="built_in">read</span> length threshold (ignore reads shorter than &lt;int&gt;)</span><br><span class="line">   -d/-m &lt;int&gt;         maximum coverage depth [8000]</span><br><span class="line">   -q &lt;int&gt;            base quality threshold</span><br><span class="line">   -Q &lt;int&gt;            mapping quality threshold</span><br><span class="line">   -r &lt;chr:from-to&gt;    region</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">The output is a simple tab-separated table with three columns: reference name,</span><br><span class="line">position, and coverage depth.  Note that positions with zero coverage may be</span><br><span class="line">omitted by default; see the -a option.</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools depth aln.sorted.bam</code></li>
</ul>
</li>
</ul>
<h4 id="flagstat"><a href="#flagstat" class="headerlink" title="flagstat"></a>flagstat</h4><ul>
<li>– simple stats 统计bam文件中read的比对情况</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flagstat</span><br><span class="line"></span><br><span class="line">Usage: samtools flagstat [--input-fmt-option OPT=VAL] &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools flagstat aln.sorted.bam</code></li>
</ul>
<h4 id="idxstats"><a href="#idxstats" class="headerlink" title="idxstats"></a>idxstats</h4><p> – BAM index stats  统计bam索引文件里的比对信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools idxstats</span><br><span class="line"></span><br><span class="line">Usage: samtools idxstats &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools idxstats aln.sorted.bam</code></li>
<li>idxstats生成一个统计表格，共有4列，分别为”序列名，序列长度，比对上的reads数，unmapped reads number”。第4列应该是paired reads中有一端能匹配到该scaffold上，而另外一端不匹配到任何scaffolds上的reads数。</li>
</ul>
<h4 id="phase"><a href="#phase" class="headerlink" title="phase"></a>phase</h4><p>– Call and phase heterozygous SNPs</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">$ samtools phase</span><br><span class="line"></span><br><span class="line">Usage:   samtools phase [options] &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line">Options: -k INT    block length [13]</span><br><span class="line">         -b STR    prefix of BAMs to output [null]</span><br><span class="line">         -q INT    min het phred-LOD [37]</span><br><span class="line">         -Q INT    min base quality <span class="keyword">in</span> het calling [13]</span><br><span class="line">         -D INT    max <span class="built_in">read</span> depth [256]</span><br><span class="line">         -F        <span class="keyword">do</span> not attempt to fix chimeras</span><br><span class="line">         -A        drop reads with ambiguous phase</span><br><span class="line"></span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools phase test.bam</code></li>
</ul>
<h4 id="stats"><a href="#stats" class="headerlink" title="stats"></a>stats</h4><ul>
<li><p>– generate stats (former bamcheck)  对bam文件做详细统计, 统计结果可用mics&#x2F;plot-bamstats作图</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">$ samtools stats</span><br><span class="line"></span><br><span class="line">About: The program collects statistics from BAM files. The output can be visualized using plot-bamstats.</span><br><span class="line">Usage: samtools stats [OPTIONS] file.bam</span><br><span class="line">       samtools stats [OPTIONS] file.bam chr:from-to</span><br><span class="line">Options:</span><br><span class="line">    -c, --coverage &lt;int&gt;,&lt;int&gt;,&lt;int&gt;    Coverage distribution min,max,step [1,1000,1]</span><br><span class="line">    -d, --remove-dups                   Exclude from statistics reads marked as duplicates</span><br><span class="line">    -f, --required-flag  &lt;str|int&gt;      Required flag, 0 <span class="keyword">for</span> <span class="built_in">unset</span>. See also `samtools flags` [0]</span><br><span class="line">    -F, --filtering-flag &lt;str|int&gt;      Filtering flag, 0 <span class="keyword">for</span> <span class="built_in">unset</span>. See also `samtools flags` [0]</span><br><span class="line">        --GC-depth &lt;<span class="built_in">float</span>&gt;              the size of GC-depth bins (decreasing bin size increases memory requirement) [2e4]</span><br><span class="line">    -h, --<span class="built_in">help</span>                          This <span class="built_in">help</span> message</span><br><span class="line">    -i, --insert-size &lt;int&gt;             Maximum insert size [8000]</span><br><span class="line">    -I, --<span class="built_in">id</span> &lt;string&gt;                   Include only listed <span class="built_in">read</span> group or sample name</span><br><span class="line">    -l, --read-length &lt;int&gt;             Include <span class="keyword">in</span> the statistics only reads with the given <span class="built_in">read</span> length []</span><br><span class="line">    -m, --most-inserts &lt;<span class="built_in">float</span>&gt;          Report only the main part of inserts [0.99]</span><br><span class="line">    -P, --split-prefix &lt;str&gt;            Path or string prefix <span class="keyword">for</span> filepaths output by -S (default is input filename)</span><br><span class="line">    -q, --trim-quality &lt;int&gt;            The BWA trimming parameter [0]</span><br><span class="line">    -r, --ref-seq &lt;file&gt;                Reference sequence (required <span class="keyword">for</span> GC-depth and mismatches-per-cycle calculation).</span><br><span class="line">    -s, --sam                           Ignored (input format is auto-detected).</span><br><span class="line">    -S, --<span class="built_in">split</span> &lt;tag&gt;                   Also write statistics to separate files <span class="built_in">split</span> by tagged field.</span><br><span class="line">    -t, --target-regions &lt;file&gt;         Do stats <span class="keyword">in</span> these regions only. Tab-delimited file chr,from,to, 1-based, inclusive.</span><br><span class="line">    -x, --sparse                        Suppress outputting IS rows <span class="built_in">where</span> there are no insertions.</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools stats aln.sorted.bam</code></li>
<li>输出的信息比较多，部分如下：<br> Summary Numbers，raw total sequences，filtered sequences, reads mapped, reads mapped and paired,reads properly paired等信息<br> Fragment Qualitites：根据cycle统计每个位点上的碱基质量分布<br> Coverage distribution：深度为1，2，3，，，的碱基数目<br> ACGT content per cycle：ACGT在每个cycle中的比例<br> Insert sizes：插入长度的统计<br> Read lengths：read的长度分布</li>
</ul>
</li>
</ul>
<h3 id="2-5-可视化-Viewing"><a href="#2-5-可视化-Viewing" class="headerlink" title="2.5 可视化(Viewing)"></a>2.5 可视化(Viewing)</h3><h4 id="flags"><a href="#flags" class="headerlink" title="flags"></a>flags</h4><ul>
<li><p>– explain BAM flags   查看不同flag值的含义</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flags</span><br><span class="line"></span><br><span class="line">About: Convert between textual and numeric flag representation</span><br><span class="line">Usage: samtools flags INT|STR[,...]</span><br><span class="line"></span><br><span class="line">Flags:</span><br><span class="line">        0x1     PAIRED        .. paired-end (or multiple-segment) sequencing technology</span><br><span class="line">        0x2     PROPER_PAIR   .. each segment properly aligned according to the aligner</span><br><span class="line">        0x4     UNMAP         .. segment unmapped</span><br><span class="line">        0x8     MUNMAP        .. next segment <span class="keyword">in</span> the template unmapped</span><br><span class="line">        0x10    REVERSE       .. SEQ is reverse complemented</span><br><span class="line">        0x20    MREVERSE      .. SEQ of the next segment <span class="keyword">in</span> the template is reversed</span><br><span class="line">        0x40    READ1         .. the first segment <span class="keyword">in</span> the template</span><br><span class="line">        0x80    READ2         .. the last segment <span class="keyword">in</span> the template</span><br><span class="line">        0x100   SECONDARY     .. secondary alignment</span><br><span class="line">        0x200   QCFAIL        .. not passing quality controls</span><br><span class="line">        0x400   DUP           .. PCR or optical duplicate</span><br><span class="line">        0x800   SUPPLEMENTARY .. supplementary alignment</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example: <code>samtools flags PAIRED,UNMAP,MUNMAP</code></p>
</li>
<li><p><strong>FLAGS:</strong></p>
<table>
<thead>
<tr>
<th>0x1</th>
<th>PAIRED</th>
<th align="center">paired-end (or multiple-segment) sequencing technology</th>
</tr>
</thead>
<tbody><tr>
<td>0x2</td>
<td>PROPER_PAIR</td>
<td align="center">each segment properly aligned according to the aligner</td>
</tr>
<tr>
<td>0x4</td>
<td>UNMAP</td>
<td align="center">segment unmapped</td>
</tr>
<tr>
<td>0x8</td>
<td>MUNMAP</td>
<td align="center">next segment in the template unmapped</td>
</tr>
<tr>
<td>0x10</td>
<td>REVERSE</td>
<td align="center">SEQ is reverse complemented</td>
</tr>
<tr>
<td>0x20</td>
<td>MREVERSE</td>
<td align="center">SEQ of the next segment in the template is reverse complemented</td>
</tr>
<tr>
<td>0x40</td>
<td>READ1</td>
<td align="center">the first segment in the template</td>
</tr>
<tr>
<td>0x80</td>
<td>READ2</td>
<td align="center">the last segment in the template</td>
</tr>
<tr>
<td>0x100</td>
<td>SECONDARY</td>
<td align="center">secondary alignment</td>
</tr>
<tr>
<td>0x200</td>
<td>QCFAIL</td>
<td align="center">not passing quality controls</td>
</tr>
<tr>
<td>0x400</td>
<td>DUP</td>
<td align="center">PCR or optical duplicate</td>
</tr>
<tr>
<td>0x800</td>
<td>SUPPLEMENTARY</td>
<td align="center">supplementary alignment</td>
</tr>
</tbody></table>
</li>
</ul>
</li>
</ul>
<h4 id="tview"><a href="#tview" class="headerlink" title="tview"></a>tview</h4><p> – text alignment viewer  直观地显示reads比对到参考基因组的情况，类似于基因组浏览器</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ samtools tview</span><br><span class="line"></span><br><span class="line">Usage: samtools tview [options] &lt;aln.bam&gt; [ref.fasta]</span><br><span class="line">Options:</span><br><span class="line">   -d display      output as (H)tml or (C)urses or (T)ext</span><br><span class="line">   -p chr:pos      go directly to this position</span><br><span class="line">   -s STR          display only reads from this sample or group</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools tview aln.sorted.bam ref.fasta</code></li>
<li>需先对bam文件进行排序并构建索引</li>
</ul>
<h4 id="view"><a href="#view" class="headerlink" title="view"></a>view</h4><p>– SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion 将sam和bam文件进行格式互换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view</span><br><span class="line"></span><br><span class="line">Usage: samtools view [options] &lt;in.bam&gt;|&lt;in.sam&gt;|&lt;in.cram&gt; [region ...]</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -b       output BAM</span><br><span class="line">  -C       output CRAM (requires -T)</span><br><span class="line">  -1       use fast BAM compression (implies -b)</span><br><span class="line">  -u       uncompressed BAM output (implies -b)</span><br><span class="line">  -h       include header <span class="keyword">in</span> SAM output</span><br><span class="line">  -H       <span class="built_in">print</span> SAM header only (no alignments)</span><br><span class="line">  -c       <span class="built_in">print</span> only the count of matching records</span><br><span class="line">  -o FILE  output file name [stdout]</span><br><span class="line">  -U FILE  output reads not selected by filters to FILE [null]</span><br><span class="line">  -t FILE  FILE listing reference names and lengths (see long <span class="built_in">help</span>) [null]</span><br><span class="line">  -L FILE  only include reads overlapping this BED FILE [null]</span><br><span class="line">  -r STR   only include reads <span class="keyword">in</span> <span class="built_in">read</span> group STR [null]</span><br><span class="line">  -R FILE  only include reads with <span class="built_in">read</span> group listed <span class="keyword">in</span> FILE [null]</span><br><span class="line">  -q INT   only include reads with mapping quality &gt;= INT [0]</span><br><span class="line">  -l STR   only include reads <span class="keyword">in</span> library STR [null]</span><br><span class="line">  -m INT   only include reads with number of CIGAR operations consuming</span><br><span class="line">           query sequence &gt;= INT [0]</span><br><span class="line">  -f INT   only include reads with all bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -F INT   only include reads with none of the bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -x STR   <span class="built_in">read</span> tag to strip (repeatable) [null]</span><br><span class="line">  -B       collapse the backward CIGAR operation</span><br><span class="line">  -s FLOAT <span class="built_in">integer</span> part sets seed of random number generator [0];</span><br><span class="line">           rest sets fraction of templates to subsample [no subsampling]</span><br><span class="line">  -@, --threads INT</span><br><span class="line">           number of BAM/CRAM compression threads [0]</span><br><span class="line">  -?       <span class="built_in">print</span> long <span class="built_in">help</span>, including note about region specification</span><br><span class="line">  -S       ignored (input format is auto-detected)</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -T, --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example ：<code>samtools view -bS test.sam &gt; test.bam</code></p>
</li>
<li><p>重要参数解读：</p>
<p>-b：输出bam格式</p>
<p>-C：输出CRAM文件</p>
<p>-1：快速压缩（需要-b）</p>
<p>-u：输出未压缩的bam文件，节约时间，占据较多磁盘空间（需要-b）</p>
<p>-h：默认输出sam文件不带表头，该参数设定后输出带表头信息</p>
<p>-H：仅仅输出表头信息</p>
<p>-c：仅打印匹配数</p>
<p>-o：输出文件（stdout标准输出）</p>
<p>-U：输出没有经过过滤选择的reads</p>
<p>-t：制表分隔符文件（需要提供额外的参考数据，比如参考基因组的索引）</p>
<p>-L：仅包括和bed文件有重叠的reads</p>
<p>-r：仅输出在STR读段组中的reads</p>
<p>-R：仅输出特定reads</p>
<p>-q：定位的质量大于INT[默认0]</p>
<p>-l：仅输出在STR 库中的reads</p>
<p>-F：获得比对上（mapped）的过滤设置[默认0]</p>
<p>-f：获得未必对上（unmapped）的过滤设置[默认0]</p>
<p>-T：使用fasta格式的参考序列</p>
</li>
</ul>
<h4 id="depad"><a href="#depad" class="headerlink" title="depad"></a>depad</h4><ul>
<li><p>– convert padded BAM to unpadded BAM</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools depad</span><br><span class="line"></span><br><span class="line">Usage:   samtools depad &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -s           Output is SAM (default is BAM)</span><br><span class="line">  -S           Input is SAM (default is BAM)</span><br><span class="line">  -u           Uncompressed BAM output (can<span class="string">&#x27;t use with -s)</span></span><br><span class="line"><span class="string">  -1           Fast compression BAM output (can&#x27;</span>t use with -s)</span><br><span class="line">  -T, --reference FILE</span><br><span class="line">               Padded reference sequence file [null]</span><br><span class="line">  -o FILE      Output file name [stdout]</span><br><span class="line">  -?           Longer <span class="built_in">help</span></span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools depad input.bam</code></li>
</ul>
</li>
</ul>
<h3 id="2-6-常用示例"><a href="#2-6-常用示例" class="headerlink" title="2.6 常用示例"></a>2.6 常用示例</h3><p>sam&#x2F;bam格式文件互换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bS -1 test.sam &gt; test.bam <span class="comment"># sam转bam</span></span><br><span class="line">$ samtools view -h test.bam &gt; test.sam <span class="comment"># bam转sam</span></span><br></pre></td></tr></table></figure>

<p>提取比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bF 4 test.bam &gt; test.F.bam</span><br></pre></td></tr></table></figure>

<p>提取没有比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bf 4 test.bam &gt; test.f.bam</span><br></pre></td></tr></table></figure>

<p>双端reads都比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bF 12 test.bam &gt; test.12.bam</span><br></pre></td></tr></table></figure>

<p>单端reads1比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -bF 4 -f 8  <span class="built_in">test</span> .bam &gt; test1.bam</span><br></pre></td></tr></table></figure>

<p>单端reads2比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bF 8 -f 4 test.bam &gt; test2.bam</span><br></pre></td></tr></table></figure>

<p>提取bam文件中比对到scaffold1上的比对结果，并保存到sam文件格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view abc.bam scaffold1 &gt; scaffold1.sam</span><br></pre></td></tr></table></figure>

<p>提取scaffold1上能比对到30k到100k区域的比对结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view abc.bam scaffold1:30000-100000 &gt; scaffold1_30k-100k.sam</span><br></pre></td></tr></table></figure>

<p>根据fasta文件，将 header 加入到 sam 或 bam 文件中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -T genome.fasta -h scaffold1.sam &gt; scaffold1.h.sam</span><br></pre></td></tr></table></figure>

<p>call SNP和INDEL等变异信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ samtools mpileup -f genome.fasta abc.bam &gt; abc.txt</span><br><span class="line">$ samtools mpileup -gSDf genome.fasta abc.bam &gt; abc.bcf</span><br><span class="line">$ samtools mpileup -guSDf genome.fasta abc.bam | \</span><br><span class="line">           bcftools view -cvNg - &gt; abc.vcf</span><br></pre></td></tr></table></figure>
      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2023/02/27/%E8%82%BF%E7%98%A4%E6%A3%80%E6%B5%8B-%E5%9F%BA%E7%A1%80%E6%A6%82%E5%BF%B5/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Chao Zhang">
      <meta itemprop="description" content="python, shell, R, perl, java">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Chaopower's Blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2023/02/27/%E8%82%BF%E7%98%A4%E6%A3%80%E6%B5%8B-%E5%9F%BA%E7%A1%80%E6%A6%82%E5%BF%B5/" class="post-title-link" itemprop="url">肿瘤检测-基础概念</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2023-02-27 21:02:58 / 修改时间：21:14:19" itemprop="dateCreated datePublished" datetime="2023-02-27T21:02:58+08:00">2023-02-27</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E6%B5%81%E7%A8%8B/" itemprop="url" rel="index"><span itemprop="name">流程</span></a>
                </span>
                  ，
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E6%B5%81%E7%A8%8B/%E8%82%BF%E7%98%A4%E6%A3%80%E6%B5%8B/" itemprop="url" rel="index"><span itemprop="name">肿瘤检测</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h1 id="1-概念汇总"><a href="#1-概念汇总" class="headerlink" title="1. 概念汇总"></a>1. 概念汇总</h1><h2 id="1-1-参考基因"><a href="#1-1-参考基因" class="headerlink" title="1.1 参考基因"></a>1.1 参考基因</h2><p>ucsc.hg19.fasta 人类基因组参考序列，最新版本是hg38，但因hg19版本配套的数据库和软件支持较全，所以临床较多使用hg19版本。</p>
<p>hg19_refGene.txt NCBI的基因参考序列数据库，提供基因的位置信息和基本的功能单元如外显子(exon)、内含子(intron)、非编码区(UTR)等的位置信息。</p>
<p>hg19_cytoBand.txt cytogenetic band 的简称，是每个细胞间细胞发生带(cytogenetic band )的染色体坐标信息。</p>
<p>1000g 千人基因组计划，包含世界几个族群的等位基因组频率，包含汉族和东亚族群的等位基因频率。</p>
<p>hg19_esp6500siv2_all.txt ESP全称是NHLBI GO(National Heart.Lung.and Blood Institule Grand Opportunity) Exome Sequencing Project（美国国家心肺和血液研究所外显子组测序计划），包含了SNP和InDel变异。目前有6503个样本，分别来源于非裔美国人2203人，欧裔美国人4300人。</p>
<p> hg19_exac03.txt 版本：03ExAC全称是The Exome Aggregation Consortium（外显子组整合数据库），整合了6万多个无亲缘关系个体的数据，这些个体来源于大量疾病研究和群体遗传学研究，能够用做严重疾病研究的参考数据库，目前EXAC数据库中包括ALL, AFR(African)，,AMR(Admixed American), EAS(East Asian), FIN(Finnish), NFE(Non-finnish European)， OTH(other)， SAS(South Asian)。该数据库旨在汇总和协调各种大规模测序项目的外显子组测序数据，通过ExAC的注释，可以了解该变异位点上突变碱基的等位基因频率，并可采用0.01的标准进行过滤，与千人基因组项目一样，有的公司会根据不同目标研究人种单独对某个子数据库进行分析，例如研究中国人，会另外在东亚EAS数据库中注释突变频率。</p>
<h2 id="1-2-数据库介绍"><a href="#1-2-数据库介绍" class="headerlink" title="1.2 数据库介绍"></a>1.2 数据库介绍</h2><p>hg19_cosmic70.txt 全称是Catalogue of Somatic Mutation in Cancer，是世界上研究人体肿瘤体细胞突变最大和最全面的数据库，癌症相关的体细胞位点，是整个网站的核心。</p>
<p>hg19_clinvar_20210724.txt 注释变异与人类疾病之间的关系，临床意义的数据来源于NCBI，Clinvar整合了dbSNP、dbVar、Pubmed、OMIM等多个数据库在遗传变异和临床表型方面的数据信息，形成一个标准的、可信的遗传变异-临床相关的数据库。</p>
<p>hg19_intervar_20180118.txt 评价错义突变有害性的数据库，评价结果可分为Pathogenic（有害）、 Likely pathogenic（可能有害）、Benign（无害）、Likely benign（可能无害）、Uncertain significance（有害性不明确）几种。</p>
<p>hg19_dbnsfp35a.txt 用于基因突变的致病性预测。</p>
<p>avsnp 用于注释变异的rs号。</p>
<p>private_databases 整理后的oncoKB、pharmGKB、clinicaltrials等，包括靶向用药信息，化疗用药信息，遗传风险数据库信息等，为了便于版本管理，以目录结构为统一数据库名称private_databases。</p>
<h2 id="1-3-其他常规文件"><a href="#1-3-其他常规文件" class="headerlink" title="1.3 其他常规文件"></a>1.3 其他常规文件</h2><p>bed文件</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2023/02/26/%E6%8F%8F%E8%BF%B0%E6%80%A7%E7%BB%9F%E8%AE%A1%E9%87%8F/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Chao Zhang">
      <meta itemprop="description" content="python, shell, R, perl, java">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Chaopower's Blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2023/02/26/%E6%8F%8F%E8%BF%B0%E6%80%A7%E7%BB%9F%E8%AE%A1%E9%87%8F/" class="post-title-link" itemprop="url">生物统计理论</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2023-02-26 20:22:39" itemprop="dateCreated datePublished" datetime="2023-02-26T20:22:39+08:00">2023-02-26</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">更新于</span>
                <time title="修改时间：2023-02-27 21:25:09" itemprop="dateModified" datetime="2023-02-27T21:25:09+08:00">2023-02-27</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%BB%9F%E8%AE%A1%E5%AD%A6/" itemprop="url" rel="index"><span itemprop="name">统计学</span></a>
                </span>
                  ，
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%BB%9F%E8%AE%A1%E5%AD%A6/%E6%8F%8F%E8%BF%B0%E6%80%A7%E7%BB%9F%E8%AE%A1%E9%87%8F/" itemprop="url" rel="index"><span itemprop="name">描述性统计量</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h2 id="1-描述性统计量"><a href="#1-描述性统计量" class="headerlink" title="1. 描述性统计量"></a>1. 描述性统计量</h2><p>（来源生信技能树）</p>
<p>为了解决某个问题，我们通常会观察一组和该问题相关的样本，利用总体中的部分样本来推断总体的情况进而得到相关结论。在通过样本推断总体前，首先需用对已有样本数据进行简单的评估和描述，针对这一需求也就引出了描述统计量这一概念。进行描述性统计时，我们最关注数据两个层面的问题：数据的集中趋势和变异分散性。</p>
<h3 id="1-1-数据的集中趋势"><a href="#1-1-数据的集中趋势" class="headerlink" title="1.1 数据的集中趋势"></a>1.1 数据的集中趋势</h3><p>均值：所有观察值的和除以观察的个数。算数平均是最自然和常用的测度，其问题在于对异常值(outliers)非常敏感。有极端值存在时，均值不能代表样本的绝大多数情况。</p>
<p>中位数：所谓中位数，是指所有样本观测值由小到大排序，位于中间的一个（样本数为奇数）或者两个数据的平均值（样本数为偶数）。</p>
<p>当数据分布对称时，中位数近似等于算数平均数；当数据正倾斜时（图像向右倾斜），中位数小于算数平均数；当数据负倾斜时（图像向左倾斜），中位数大于算数平均数。因此，我们可以通过比较样本的均值和中位数对数据的分布对称性进行初判断。</p>
<p>众数：在样本所有观测值中，出现频率最大（出现次数最多）的数值称为众数。这里需要说明，当数据量很大而且数值不会多次重复出现时众数并不能带来太多信息。比如当计算上万个基因的表达量后，得到的众数最可能是0，因为每个基因的表达值或多或少都有一些不同，这时候出现最多的就是那些没有检测到表达基因的0了。但是在遇到类别数据而非数值型数据时众数有很大用处，或者说众数是唯一可以用于类别数据的平均数。</p>
<p>在R中，均值和中位数可以通过mean()和median()进行计算,而众数可以通过modeest包mfv()函数得到。</p>
<h3 id="1-2-数据的变异性（离散性）"><a href="#1-2-数据的变异性（离散性）" class="headerlink" title="1.2 数据的变异性（离散性）"></a>1.2 数据的变异性（离散性）</h3><p>平均数显然不能说明一切问题，在说明样本数据时我们还必须考虑数据是不是过于分散。例如在篮球队员投篮平均得分相同的情况下，更重要的是知道他们谁发挥更加稳定。</p>
<p>极差(range)指一个样本中最大值和最小值之间的差值。在统计学中也称为全距，它能够指出数据的“宽度”（范围）。但它和均值一样易受极端值影响，而且也会受样本量明显影响。</p>
<p>针对极差的缺点，统计学又引入分位数(quantiles)概念，通俗讲是把数据的“宽度”细分后再去进行比较从而更好地描述数据的分布形态。分位数用三个点将从小到大排列好的数据分为四个相等部分，而这三个点也就是我们常说的四分位数，分别叫做下四分位数，中位数和上四分位数。当然，除了四分位也可以计算十分位或者百分位。</p>
<p>分位数的引进能够说明数值的位置，但无法说明某数值在该位置出现的概率。为了说明数据的稳定程度，我们可以考虑计算每个数据值到平均数的距离（此处可以脑补一个高瘦形的数据曲线和矮胖形的数据曲线），但样本中所有观测值与均值偏差的和永远是0。为了解决这种正负距离相互抵消的问题，统计学又引入方差(variance)和标准差(standard deviation)概念。</p>
<p>所谓方差指数值与均值距离平方数的平均数，而标准差，则是方差的平方根。标准差体现了数据的变异度，标准差越小，数值和均值越近。通常均值用μ表示，而标准差用σ表示。</p>
<p>在R中，可以通过quantile()计算分位数，通过var()来计算方差，通过sd()来计算标准差。</p>
<p>有了标准差的概念，随之而来的问题是当两个样本标准差相同但是均值相差很大时该如何做出区分。于是，统计学引入了变异系数(coefficient of variation, CV) 概念，变异系数是指样本标准差除以均值再乘100%。变异系数不会受数据尺度的影响，因此常用来进行不同样本之间变异性的比较。</p>
<p>在实际的数据分析中，如果要比较不同数据集(均值和标准差都不同)之间的数值，通常会引入z score的概念，z score 的计算方法是用某一数值减去均值在除以标准差。通过对原始数据进行z变换，我们将不同数据集转化为一个新的均值为0，标准差为1的分布。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2023/02/25/%E6%A6%82%E7%8E%87%E7%9B%B8%E5%85%B3/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Chao Zhang">
      <meta itemprop="description" content="python, shell, R, perl, java">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Chaopower's Blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2023/02/25/%E6%A6%82%E7%8E%87%E7%9B%B8%E5%85%B3/" class="post-title-link" itemprop="url">概率相关</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2023-02-25 20:49:51" itemprop="dateCreated datePublished" datetime="2023-02-25T20:49:51+08:00">2023-02-25</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">更新于</span>
                <time title="修改时间：2023-02-27 21:25:09" itemprop="dateModified" datetime="2023-02-27T21:25:09+08:00">2023-02-27</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%BB%9F%E8%AE%A1%E5%AD%A6/" itemprop="url" rel="index"><span itemprop="name">统计学</span></a>
                </span>
                  ，
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%BB%9F%E8%AE%A1%E5%AD%A6/%E6%A6%82%E7%8E%87%E7%9B%B8%E5%85%B3/" itemprop="url" rel="index"><span itemprop="name">概率相关</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h3 id="1-概念"><a href="#1-概念" class="headerlink" title="1. 概念"></a>1. 概念</h3><p>样本空间(sample space)是所有可能结果的一个集合，事件(event)则是样本空间中所有感兴趣结果的一个子集。某事件的概率(probability)是该事件在无限次试验次数中的相对频率。</p>
<p>事件的交集并集和补集，概率的加法法则和乘法法则在本章中不再做介绍。</p>
<p>条件概率(Conditional Probability)用来描述与其他事件的发生相关的某事件的概率，通常被描述成“在A事件下发生事件B的概率”。在书写时用|表示，例如P(A|B)是在B发生的情况下A发生的概率。计算方式为AB同时发生的次数除以所有B发生的次数。</p>
<p>条件概率计算方法：</p>
<p><em>P</em>(<em>A</em>∣<em>B</em>)&#x3D;<em>P</em>(<em>A</em>∩<em>B</em>)&#x2F;<em>P</em>(<em>B</em>)</p>
<p>可推出: </p>
<p><em>P</em>(<em>B</em>∣<em>A</em>)&#x3D;<em>P</em>(<em>A</em>∩<em>B</em>)&#x2F;<em>P</em>(<em>A</em>) &#x3D;&gt; <em>P</em>(<em>A</em>∩<em>B</em>)&#x3D;<em>P</em>(<em>A</em>)×<em>P</em>(<em>B</em>∣<em>A</em>)</p>
<p>全概率公式：</p>
<p><em>P</em>(<em>B</em>)&#x3D;<em>P</em>(<em>A</em>)×<em>P</em>(<em>B</em>∣<em>A</em>)+<em>P</em>(<em>A</em>′)×<em>P</em>(<em>B</em>∣<em>A</em>′)</p>
<p> （贝叶斯定理分母部分）</p>
<p>贝叶斯定理：</p>
<p><em>P</em>(<em>A</em>∣<em>B</em>)&#x3D;<em>P</em>(<em>A</em>)×<em>P</em>(<em>B</em>∣<em>A</em>)+<em>P</em>(<em>A</em>′)×<em>P</em>(<em>B</em>∣<em>A</em>′)<em>P</em>(<em>A</em>)×<em>P</em>(<em>B</em>&#x2F;<em>A</em>)</p>
<p>得到的贝叶斯定理可以帮助我们计算逆条件概率。</p>
<p>在实际的生物学数据处理的过程中，我们还会接触到灵敏度(sensitivity)、特异度(specificity)、假阳性(false  positive)和假阴性(false negative) 几个概念。在疾病相关研究中，对于某一个症状，灵敏度指发病后出现症状的概率，特异度是不发病时不出现症状的概率。假阳性是指实验结果阳性但是实际为阴性，假阴性是指实验结果为阴性但是实际为阳性。基于灵敏度和特异度可以使用ROC曲线，通常来说在两个检验中，曲线下面积大的较好。</p>
<p>概率能够告诉我们事件发生的可能性，但如果想要利用概率预测未来的结果并且评估预测的确定性就需要引入概率分布。</p>
<h3 id="离散概率分布"><a href="#离散概率分布" class="headerlink" title="离散概率分布"></a>离散概率分布</h3><p>随机变量： 样本空间中，对不同事件指定有相应概率的数值函数。</p>
<p>随机变量是可以等于一系列数值的 变量，这些值都和一个特定概率关联。它的写法是<em>P</em>(<em>X</em>&#x3D;<em>x</em>)，表示随机变量X取特定值为x时的概率。随机变量包括离散和连续两种形式，所谓离散指变量只能取一些确定值。连续指的是有无限多种可能取值。</p>
<p>概率分布 也叫概率质量分布，</p>
<p><em>P</em>(<em>X</em>&#x3D;<em>x</em>)。在描述统计量中，样本的频数分布描述每个取值及对应发生次数，如果样本总数除以对应发生次数，得到的频率分布就类似于这里的概率分布。后面会提到的“拟合优度检验”就是比较有限样本频率分布和概率分布的差异。</p>
<p>如果将随机变量和样本对应起来理解，样本中均值的概念在总体（随机变量）中称为</p>
<p>期望，也叫作总体均值，表示为<em>μ</em>（和均值一致）或者E*(<em>X</em>)。计算方式是将每个可能值和概率相乘再把所有乘积相加。和均值类似，这里的期望也无法描述相关数值分散程度。</p>
<p>同样，在随机变量中也有类似于样本方差的概念，称为</p>
<p>总体方差（随机变量方差），用来表示分散程度。其计算公式为 </p>
<p><em>Va**r</em>(<em>X</em>)&#x3D;<em>E</em>(<em>X</em>−<em>μ</em>)2。而概率分布的标准差<em>σ</em>同样是方差的平方根。</p>
<p>计算<em>E</em>(<em>X</em>−<em>μ</em>)2时，首先计算每个数值x的(<em>x</em>−<em>μ</em>)2,然后再将结果乘以概率，最后把所有结果相加。</p>
<p>在数据集中，方差和标准差表示的是数据和均值的距离，在概率分布中表示特定数值概率的分散情况。方差越小，结果越接近期望。</p>
<p>累加分布函数(cumulative-distribution function, cdf): 随机变量X，对于X的任一指定值x，概率值</p>
<p><em>P</em>(<em>X</em>≤<em>x</em>)。即随机变量取值不大于指定值的概率。记作<em>F</em>(<em>x</em>)</p>
<p> 常见的离散概率分布 </p>
<p>几何分布：进行一组相互独立实验，每次实验有成功失败两种可能且每次试验概率相等，想知道第一次成功需要进行的试验次数。</p>
<p><em>P</em>(<em>X</em>&#x3D;<em>r</em>)&#x3D;<em>p<strong>q</strong>r</em>−1</p>
<p><em>P</em>(<em>X</em>&gt;<em>r</em>)&#x3D;<em>q**r</em></p>
<p><em>P</em>(<em>X</em>≤<em>r</em>)&#x3D;1−<em>q**r</em></p>
<p>期望<em>E</em>(<em>X</em>)&#x3D;<em>x</em>1；</p>
<p>方差 <em>Va**r</em>(<em>X</em>)&#x3D;<em>p</em>2<em>q</em></p>
<p>二项分布：进行一组相互独立试验，每次实验有成功失败两种可能，每次试验概率相等且试验次数有限，想知道在有限次试验中成功的次数。</p>
<p><em>P</em>(<em>X</em>&#x3D;<em>r</em>)&#x3D;<em>C<strong>n</strong>r<strong>p</strong>r<strong>q</strong>n</em>−<em>r</em></p>
<p>期望<em>E</em>(<em>X</em>)&#x3D;<em>n**p</em>；方差<em>Va**r</em>(<em>X</em>)&#x3D;<em>n**pq</em></p>
<p>二项分布和几何分布差别在于，前者试验次数固定求成功概率，后者求第一次成功前试验次数。</p>
<p>泊松分布</p>
<p>：常与稀有事件相关，单独事件在给定区间（区间可以是时间或者空间）内随机独立发生，该区间内的事件平均发生次数已知且为有限值。这个值用<em>λ</em>表示。给定区间内发生r次事件的概率计算公式：</p>
<p><em>P</em>(<em>X</em>&#x3D;<em>r</em>)&#x3D;<em>r</em>!<em>e</em>−<em>λ<strong>λ</strong>r</em></p>
<p>期望是给定区间内能够期望的事件发生次数λ，方差也是λ。如果一个离散随机变量的一批数据计算后方差和均值近似相等，则可以推测样本符合泊松分布。当二项分布的p很小且n非常大时，<em>n**pq</em>≈<em>n**p</em></p>
<p>,方差和期望近似相等，可以用泊松分布来近似二项分布，从而使计算简化。通常，n大于50且p&lt;0.1时为典型的近似情况。</p>
<p> 连续概率分布 </p>
<p>当数据连续分布，人们关心的是取得某一个特定范围的概率。</p>
<p>概率密度函数(probability density function, pdf):本质是一个函数，用这个函数可以求出在一个范围内某连续变量的概率，同时该函数可以指出该概率分布的形状。换句话讲，任意a,b两点之间及函数对应曲线下组成的面积等于随机变量X落在ab间的概率。曲线下面积总和是1。</p>
<p>概率密度可以指出各种范围内的概率大小，用面积来表示。概率密度只是表示概率的一种方法而非概率本身。</p>
<p>累加分布函数</p>
<p>a点上的值等于随机变量X取值≤<em>a</em>的概率，也是概率密度函数a左边曲线下的面积。</p>
<p>正态分布</p>
<p>是连续数据的一种理想状态。正态分布的概率密度函数是一条对称的钟形曲线，均值具有最大的概率密度，偏离均值概率密度逐渐变小。参数<em>μ</em>是均值，也是曲线的中央位置，<em>σ</em>表述曲线的“胖瘦”。</p>
<p>连续随机变量X符合均值为<em>μ</em>，标准差为<em>σ</em>的正态分布时写作<em>X</em>∼<em>N</em>(<em>μ</em>,<em>σ</em>2).</p>
<p>线性变换</p>
<p><em>a**X</em>+<em>b</em>∼<em>N</em>(<em>a**μ</em>+<em>b</em>,<em>a</em>2<em>σ</em>2)</p>
<p>当X和Y相互独立（彼此之间没有影响）时：</p>
<p><em>X</em>+<em>Y</em>∼<em>N</em>(<em>μ**x</em>+<em>μ**y</em>,<em>σ**x</em>2+<em>σ**y</em>2)</p>
<p><em>X</em>−<em>Y</em>∼<em>N</em>(<em>μ**x</em>−<em>μ**y</em>,<em>σ**x</em>2+<em>σ**y</em>2)</p>
<p>期望</p>
<p><em>E</em>(<em>X</em>1+<em>X</em>2+…+<em>X**n</em>)&#x3D;<em>n**E</em>(<em>X</em>)</p>
<p>；方差<em>Va**r</em>(<em>X</em>1+<em>X</em>2+…+<em>X**n</em>)&#x3D;<em>nVa**r</em>(<em>X</em>)</p>
<p>当X和Y并不彼此独立时，使用协方差(Covariance)来描述两个随机变量间的关系，记做Cov(X,Y)</p>
<p><em>C<strong>o</strong>v</em>(<em>X</em>,<em>Y</em>)&#x3D;<em>E</em>[(<em>X</em>−<em>μ**x</em>)(<em>Y</em>−<em>μ**y</em>)]二项分布正态近似</p>
<p>：通常情况，当二项分布满足</p>
<p><em>n**p</em>≥5 (也有建议<em>n**pq</em>≥5)时，可以用正态分布代替二项分布。其中<em>μ</em>&#x3D;<em>n**p</em>,<em>σ</em>2&#x3D;<em>n**pq</em>。另外，当泊松分布的λ&gt;15时，也可以用正态分布进行近似。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Chao Zhang"
      src="/images/avatar.png">
  <p class="site-author-name" itemprop="name">Chao Zhang</p>
  <div class="site-description" itemprop="description">python, shell, R, perl, java</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">5</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">6</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">标签</span>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/chaopower-zhang" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;chaopower-zhang" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="/chaopower@126.com" title="E-Mail → chaopower@126.com"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>
  <div class="cc-license motion-element" itemprop="license">
    <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" class="cc-opacity" rel="noopener" target="_blank"><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"></a>
  </div>



      </div>
        <div class="back-to-top motion-element">
          <i class="fa fa-arrow-up"></i>
          <span>0%</span>
        </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">chaopower@126.com</span>
</div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

</body>
</html>
