<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/C.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/C.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":true,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="samtools 常用命令1.软件介绍及安装Samtools 用于处理sam与bam格式的工具软件，能够实现二进制查看、格式转换排序及合并等功能，结合sam格式中的flag、tag等信息，还可以完成比对结果的统计汇总，同时利用linux中的grep、awk等操作命令，还可以大大扩展samtools的使用范围与功能。 123456789101112131415#获取压缩包并解压wgethttps:&#x2F;">
<meta property="og:type" content="article">
<meta property="og:title" content="生信软件-samtools">
<meta property="og:url" content="http://example.com/2023/03/01/samtools%20%E4%BD%BF%E7%94%A8/index.html">
<meta property="og:site_name" content="Chaopower&#39;s Blog">
<meta property="og:description" content="samtools 常用命令1.软件介绍及安装Samtools 用于处理sam与bam格式的工具软件，能够实现二进制查看、格式转换排序及合并等功能，结合sam格式中的flag、tag等信息，还可以完成比对结果的统计汇总，同时利用linux中的grep、awk等操作命令，还可以大大扩展samtools的使用范围与功能。 123456789101112131415#获取压缩包并解压wgethttps:&#x2F;">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2023-03-01T13:02:58.000Z">
<meta property="article:modified_time" content="2023-03-01T11:10:14.906Z">
<meta property="article:author" content="Chao Zhang">
<meta property="article:tag" content="工具">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/2023/03/01/samtools%20%E4%BD%BF%E7%94%A8/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'zh-CN'
  };
</script>

  <title>生信软件-samtools | Chaopower's Blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Chaopower's Blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">chaopower@126.com</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>关于</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>标签<span class="badge">3</span></a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>分类<span class="badge">6</span></a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>归档<span class="badge">4</span></a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2023/03/01/samtools%20%E4%BD%BF%E7%94%A8/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Chao Zhang">
      <meta itemprop="description" content="python, shell, R, perl, java">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Chaopower's Blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          生信软件-samtools
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2023-03-01 21:02:58 / 修改时间：19:10:14" itemprop="dateCreated datePublished" datetime="2023-03-01T21:02:58+08:00">2023-03-01</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E7%94%9F%E4%BF%A1%E8%BD%AF%E4%BB%B6/" itemprop="url" rel="index"><span itemprop="name">生信软件</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h1 id="samtools-常用命令"><a href="#samtools-常用命令" class="headerlink" title="samtools 常用命令"></a>samtools 常用命令</h1><h2 id="1-软件介绍及安装"><a href="#1-软件介绍及安装" class="headerlink" title="1.软件介绍及安装"></a>1.软件介绍及安装</h2><p>Samtools 用于处理sam与bam格式的工具软件，能够实现二进制查看、格式转换排序及合并等功能，结合sam格式中的flag、tag等信息，还可以完成比对结果的统计汇总，同时利用linux中的grep、awk等操作命令，还可以大大扩展samtools的使用范围与功能。</p>
<figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#获取压缩包并解压</span></span><br><span class="line">wget</span><br><span class="line">https://github.com/samtools/samtools/releases/download/1.9/samtools</span><br><span class="line">1.9.tar.bz2</span><br><span class="line">tar -xzvf samtools-1.9.tar .bz2</span><br><span class="line"><span class="comment">#进入samtools文件夹</span></span><br><span class="line"><span class="built_in">cd</span> samtools-1.9</span><br><span class="line"><span class="comment">#安装</span></span><br><span class="line">./configure</span><br><span class="line">make</span><br><span class="line">sudo make install</span><br><span class="line"><span class="comment">##非root安装</span></span><br><span class="line"><span class="comment">#./configure --prefix=/home/用户名/.Local</span></span><br><span class="line"><span class="comment">#make</span></span><br><span class="line"><span class="comment">#make install</span></span><br></pre></td></tr></table></figure>

<h2 id="2-Samtools-基本命令"><a href="#2-Samtools-基本命令" class="headerlink" title="2.Samtools 基本命令"></a>2.Samtools 基本命令</h2><figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line">$ samtools --<span class="built_in">help</span></span><br><span class="line"></span><br><span class="line">Program: samtools (Tools <span class="keyword">for</span> alignments <span class="keyword">in</span> the SAM format)</span><br><span class="line">Version: 1.3.1 (using htslib 1.3.1)</span><br><span class="line"></span><br><span class="line">Usage:   samtools &lt;<span class="built_in">command</span>&gt; [options]</span><br><span class="line"></span><br><span class="line">Commands:</span><br><span class="line">  -- Indexing</span><br><span class="line">     dict           create a sequence dictionary file</span><br><span class="line">     faidx          index/extract FASTA</span><br><span class="line">     index          index alignment</span><br><span class="line"></span><br><span class="line">  -- Editing</span><br><span class="line">     calmd          recalculate MD/NM tags and <span class="string">&#x27;=&#x27;</span> bases</span><br><span class="line">     fixmate        fix mate information</span><br><span class="line">     reheader       replace BAM header</span><br><span class="line">     rmdup          remove PCR duplicates</span><br><span class="line">     targetcut      <span class="built_in">cut</span> fosmid regions (<span class="keyword">for</span> fosmid pool only)</span><br><span class="line">     addreplacerg   adds or replaces RG tags</span><br><span class="line"></span><br><span class="line">  -- File operations</span><br><span class="line">     collate        shuffle and group alignments by name</span><br><span class="line">     <span class="built_in">cat</span>            concatenate BAMs</span><br><span class="line">     merge          merge sorted alignments</span><br><span class="line">     mpileup        multi-way pileup</span><br><span class="line">     <span class="built_in">sort</span>           <span class="built_in">sort</span> alignment file</span><br><span class="line">     <span class="built_in">split</span>          splits a file by <span class="built_in">read</span> group</span><br><span class="line">     quickcheck     quickly check <span class="keyword">if</span> SAM/BAM/CRAM file appears intact</span><br><span class="line">     fastq          converts a BAM to a FASTQ</span><br><span class="line">     fasta          converts a BAM to a FASTA</span><br><span class="line"></span><br><span class="line">  -- Statistics</span><br><span class="line">     bedcov         <span class="built_in">read</span> depth per BED region</span><br><span class="line">     depth          compute the depth</span><br><span class="line">     flagstat       simple stats</span><br><span class="line">     idxstats       BAM index stats</span><br><span class="line">     phase          phase heterozygotes</span><br><span class="line">     stats          generate stats (former bamcheck)</span><br><span class="line"></span><br><span class="line">  -- Viewing</span><br><span class="line">     flags          explain BAM flags</span><br><span class="line">     tview          text alignment viewer</span><br><span class="line">     view           SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion</span><br><span class="line">     depad          convert padded BAM to unpadded BAM</span><br></pre></td></tr></table></figure>

<h3 id="2-1-构建索引"><a href="#2-1-构建索引" class="headerlink" title="2.1 构建索引"></a>2.1 构建索引</h3><h4 id="2-1-1-dict"><a href="#2-1-1-dict" class="headerlink" title="2.1.1 dict"></a>2.1.1 dict</h4><p><strong>dict</strong> – 对fasta格式的参考基因组序列构建字典索引</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ samtools dict</span><br><span class="line"></span><br><span class="line">About:   Create a sequence dictionary file from a fasta file</span><br><span class="line">Usage:   samtools dict [options] &lt;file.fa|file.fa.gz&gt;</span><br><span class="line"></span><br><span class="line">Options: -a, --assembly STR    assembly</span><br><span class="line">         -H, --no-header       do not print @HD line</span><br><span class="line">         -o, --output STR      file to write out dict file [stdout]</span><br><span class="line">         -s, --species STR     species</span><br><span class="line">         -u, --uri STR         URI [file:///abs/path/to/file.fa]</span><br></pre></td></tr></table></figure>

<p>Example：<code>samtools dict -a GRCh38 -s &quot;Homo sapiens&quot; hg38.fasta -o hg38.dict</code></p>
<p>构建完成后会生成一个hg38.dict的参考基因组索引文件</p>
<h4 id="2-1-2-faidx"><a href="#2-1-2-faidx" class="headerlink" title="2.1.2  faidx"></a>2.1.2  faidx</h4><p><strong>faidx</strong> – 对fasta格式的文件建立索引，生成的索引文件以.fai后缀结尾。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools faidx</span><br><span class="line"></span><br><span class="line">Usage:   samtools faidx &lt;file.fa|file.fa.gz&gt; [&lt;reg&gt; [...]]</span><br></pre></td></tr></table></figure>

<p>Example：<code>samtools faidx hg38.fasta</code></p>
<p>构建完成后会生成一个hg38.fasta.fai的索引文件，该文件共有五列，分别表示：</p>
<p>第一列：序列的名称</p>
<p>第二列：序列的长度</p>
<p>第三列：第一个碱基的偏移量，从0开始计数</p>
<p>第四列：除了最后一行外，序列中每行的碱基数</p>
<p>第五列：除了最后一行外，序列中每行的长度（包括换行符）</p>
<p><strong>该命令也可以根据索引文件，快速提取fasta文件中任意区域的序列文件</strong></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># 提取1号染色体的序列信息</span><br><span class="line">$ samtools faidx ha38.fasta chr1 &gt; hg38.chr1.fa</span><br><span class="line"></span><br><span class="line"># 提取1号染色体200-1000bp之间的序列</span><br><span class="line">$ samtools faidx hg38.fasta chr1:200-1000 &gt; hg38.chr1_200_1000.fa</span><br></pre></td></tr></table></figure>

<h4 id="2-1-3-index"><a href="#2-1-3-index" class="headerlink" title="2.1.3 index"></a>2.1.3 index</h4><p> index – 对bam格式的比对文件构建索引，需对bam文件进行排序后才能构建索引</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ samtools index</span><br><span class="line"></span><br><span class="line">Usage: samtools index [-bc] [-m INT] &lt;in.bam&gt; [out.index]</span><br><span class="line">Options:</span><br><span class="line">  -b       Generate BAI-format index for BAM files [default]</span><br><span class="line">  -c       Generate CSI-format index for BAM files</span><br><span class="line">  -m INT   Set minimum interval size for CSI indices to 2^INT [14]</span><br></pre></td></tr></table></figure>

<p>Example： <code>samtools index aln.sorted.bam</code></p>
<p>构建完成后将产生后缀为.bai的文件，用于快速的随机处理</p>
<h3 id="2-2-文件编辑-Editing"><a href="#2-2-文件编辑-Editing" class="headerlink" title="2.2 文件编辑(Editing)"></a>2.2 文件编辑(Editing)</h3><h4 id="calmd"><a href="#calmd" class="headerlink" title="calmd"></a>calmd</h4><ul>
<li>– recalculate MD&#x2F;NM tags and ‘&#x3D;’ bases 计算MD tag标签值（ 记录了mismatch信息 ）</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools calmd</span><br><span class="line">Usage: samtools calmd [-eubrAES] &lt;aln.bam&gt; &lt;ref.fasta&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -e       change identical bases to <span class="string">&#x27;=&#x27;</span></span><br><span class="line">  -u       uncompressed BAM output (<span class="keyword">for</span> piping)</span><br><span class="line">  -b       compressed BAM output</span><br><span class="line">  -S       ignored (input format is auto-detected)</span><br><span class="line">  -A       modify the quality string</span><br><span class="line">  -r       compute the BQ tag (without -A) or <span class="built_in">cap</span> baseQ by BAQ (with -A)</span><br><span class="line">  -E       extended BAQ <span class="keyword">for</span> better sensitivity but lower specificity</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example：<code>samtools calmd -AEur aln.bam ref.fa</code></li>
</ul>
<h4 id="fixmate"><a href="#fixmate" class="headerlink" title="fixmate"></a>fixmate</h4><ul>
<li>– fix mate information  为以名称排序定位的alignment填入配对坐标，ISIZE（inferred insert size推测的插入序列大小）和配对相应的标签（flag）</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools fixmate</span><br><span class="line"></span><br><span class="line">Usage: samtools fixmate &lt;in.nameSrt.bam&gt; &lt;out.nameSrt.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -r           Remove unmapped reads and secondary alignments</span><br><span class="line">  -p           Disable FR proper pair check</span><br><span class="line">  -c           Add template cigar ct tag</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">As elsewhere <span class="keyword">in</span> samtools, use <span class="string">&#x27;-&#x27;</span> as the filename <span class="keyword">for</span> stdin/stdout. The input</span><br><span class="line">file must be grouped by <span class="built_in">read</span> name (e.g. sorted by name). Coordinated sorted</span><br><span class="line">input is not accepted.</span><br></pre></td></tr></table></figure>

<ul>
<li>Example：<code>samtools fixmate in.namesorted.sam out.bam</code></li>
</ul>
<h4 id="reheader"><a href="#reheader" class="headerlink" title="reheader"></a>reheader</h4><p> – replace BAM header 替换bam文件头注释信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">$ samtools reheader</span><br><span class="line"></span><br><span class="line">Usage: samtools reheader [-P] in.header.sam in.bam &gt; out.bam</span><br><span class="line">   or  samtools reheader [-P] -i in.header.sam file.bam</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">    -P, --no-PG      Do not generate an @PG header line.</span><br><span class="line">    -i, --in-place   Modify the bam/cram file directly.</span><br><span class="line">                     (Defaults to outputting to stdout.)</span><br></pre></td></tr></table></figure>

<h4 id="rmdup"><a href="#rmdup" class="headerlink" title="rmdup"></a>rmdup</h4><ul>
<li>– remove PCR duplicates 去除bam文件中的PCR重复信息</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">$ samtools rmdup</span><br><span class="line"></span><br><span class="line">Usage:  samtools rmdup [-sS] &lt;input.srt.bam&gt; &lt;output.bam&gt;</span><br><span class="line"></span><br><span class="line">Option: -s    rmdup <span class="keyword">for</span> SE reads</span><br><span class="line">        -S    treat PE reads as SE <span class="keyword">in</span> rmdup (force -s)</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example：<code>samtools rmdup -sS in.sorted.bam output.bam</code></li>
</ul>
<h4 id="targetcut"><a href="#targetcut" class="headerlink" title="targetcut"></a>targetcut</h4><ul>
<li><p>– cut fosmid regions (for fosmid pool only)   此命令仅用于从fosmid混池测序中切除fosmid克隆序列</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">$ samtools targetcut</span><br><span class="line"></span><br><span class="line">Usage: samtools targetcut [-Q minQ] [-i inPen] [-0 em0] [-1 em1] [-2 em2] &lt;in.bam&gt;</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -f, --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure></li>
</ul>
<h4 id="addreplacerg"><a href="#addreplacerg" class="headerlink" title="addreplacerg"></a>addreplacerg</h4><p>– adds or replaces RG tags 该命令用于添加或更改bam文件中的RG tag标签值</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools addreplacerg</span><br><span class="line"></span><br><span class="line">Usage: samtools addreplacerg [options] [-r &lt;@RG line&gt; | -R &lt;existing <span class="built_in">id</span>&gt;] [-o &lt;output.bam&gt;] &lt;input.bam&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -m MODE   Set the mode of operation from one of overwrite_all, orphan_only [overwrite_all]</span><br><span class="line">  -o FILE   Where to write output to [stdout]</span><br><span class="line">  -r STRING @RG line text</span><br><span class="line">  -R STRING ID of @RG line <span class="keyword">in</span> existing header to use</span><br><span class="line">      --input-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify input format (SAM, BAM, CRAM)</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example ：<code>samtools addreplacerg -r &#39;ID:fish&#39; -r &#39;LB:1334&#39; -r &#39;SM:alpha&#39; -o output.bam input.bam</code></li>
</ul>
<h3 id="2-3-文件处理-File-operations"><a href="#2-3-文件处理-File-operations" class="headerlink" title="2.3 文件处理(File operations)"></a>2.3 文件处理(File operations)</h3><h4 id="collate"><a href="#collate" class="headerlink" title="collate"></a>collate</h4><p>– shuffle and group alignments by name 根据read name信息将bam文件进行随机打散和分组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">$ samtools collate</span><br><span class="line"></span><br><span class="line">Usage:   samtools collate [-Ou] [-n nFiles] [-c cLevel] &lt;in.bam&gt; &lt;out.prefix&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">      -O       output to stdout</span><br><span class="line">      -u       uncompressed BAM output</span><br><span class="line">      -l INT   compression level [1]</span><br><span class="line">      -n INT   number of temporary files [64]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools collate -o aln.name_collated.bam aln.sorted.bam</code></li>
</ul>
<h4 id="cat"><a href="#cat" class="headerlink" title="cat"></a>cat</h4><p> – concatenate BAMs  将多个bam文件合并为一个bam文件（与merge命令的区别是cat不需要将bam文件提前进行排序）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools <span class="built_in">cat</span></span><br><span class="line"></span><br><span class="line">Usage: samtools <span class="built_in">cat</span> [-h header.sam] [-o out.bam] &lt;in1.bam&gt; [...]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools cat -o out.bam in1.bam in2.bam</code></li>
</ul>
<h4 id="merge"><a href="#merge" class="headerlink" title="merge"></a>merge</h4><ul>
<li>– merge sorted alignments 将多个已经sort的bam文件进行合并</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">$ samtools merge</span><br><span class="line"></span><br><span class="line">Usage: samtools merge [-nurlf] [-h inh.sam] [-b &lt;bamlist.fofn&gt;] &lt;out.bam&gt; &lt;in1.bam&gt; [&lt;in2.bam&gt; ... &lt;inN.bam&gt;]</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -n         Input files are sorted by <span class="built_in">read</span> name</span><br><span class="line">  -r         Attach RG tag (inferred from file names)</span><br><span class="line">  -u         Uncompressed BAM output</span><br><span class="line">  -f         Overwrite the output BAM <span class="keyword">if</span> exist</span><br><span class="line">  -1         Compress level 1</span><br><span class="line">  -l INT     Compression level, from 0 to 9 [-1]</span><br><span class="line">  -R STR     Merge file <span class="keyword">in</span> the specified region STR [all]</span><br><span class="line">  -h FILE    Copy the header <span class="keyword">in</span> FILE to &lt;out.bam&gt; [in1.bam]</span><br><span class="line">  -c         Combine @RG headers with colliding IDs [alter IDs to be distinct]</span><br><span class="line">  -p         Combine @PG headers with colliding IDs [alter IDs to be distinct]</span><br><span class="line">  -s VALUE   Override random seed</span><br><span class="line">  -b FILE    List of input BAM filenames, one per line [null]</span><br><span class="line">  -@, --threads INT</span><br><span class="line">             Number of BAM/CRAM compression threads [0]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools merge merged.bam in1.sorted.bam in2.sorted.bam</code></li>
</ul>
<h4 id="mpileup"><a href="#mpileup" class="headerlink" title="mpileup"></a>mpileup</h4><p> – multi-way pileup  对参考基因组每个位点做碱基堆积，用于call SNP和INDEL。主要是生成BCF、VCF文件或者pileup一个或多个bam文件。比对记录以在@RG中的样本名作为区分标识符。如果样本标识符缺失，那么每一个输入文件则视为一个样本</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br></pre></td><td class="code"><pre><span class="line">$ samtools mpileup</span><br><span class="line"></span><br><span class="line">Usage: samtools mpileup [options] in1.bam [in2.bam [...]]</span><br><span class="line"></span><br><span class="line">Input options:</span><br><span class="line">  -6, --illumina1.3+      quality is <span class="keyword">in</span> the Illumina-1.3+ encoding</span><br><span class="line">  -A, --count-orphans     <span class="keyword">do</span> not discard anomalous <span class="built_in">read</span> pairs</span><br><span class="line">  -b, --bam-list FILE     list of input BAM filenames, one per line</span><br><span class="line">  -B, --no-BAQ            <span class="built_in">disable</span> BAQ (per-Base Alignment Quality)</span><br><span class="line">  -C, --adjust-MQ INT     adjust mapping quality; recommended:50, <span class="built_in">disable</span>:0 [0]</span><br><span class="line">  -d, --max-depth INT     max per-file depth; avoids excessive memory usage [250]</span><br><span class="line">  -E, --redo-BAQ          recalculate BAQ on the fly, ignore existing BQs</span><br><span class="line">  -f, --fasta-ref FILE    faidx indexed reference sequence file</span><br><span class="line">  -G, --exclude-RG FILE   exclude <span class="built_in">read</span> <span class="built_in">groups</span> listed <span class="keyword">in</span> FILE</span><br><span class="line">  -l, --positions FILE    skip unlisted positions (chr pos) or regions (BED)</span><br><span class="line">  -q, --min-MQ INT        skip alignments with mapQ smaller than INT [0]</span><br><span class="line">  -Q, --min-BQ INT        skip bases with baseQ/BAQ smaller than INT [13]</span><br><span class="line">  -r, --region REG        region <span class="keyword">in</span> <span class="built_in">which</span> pileup is generated</span><br><span class="line">  -R, --ignore-RG         ignore RG tags (one BAM = one sample)</span><br><span class="line">  --rf, --incl-flags STR|INT  required flags: skip reads with mask bits <span class="built_in">unset</span> []</span><br><span class="line">  --ff, --excl-flags STR|INT  filter flags: skip reads with mask bits <span class="built_in">set</span></span><br><span class="line">                                            [UNMAP,SECONDARY,QCFAIL,DUP]</span><br><span class="line">  -x, --ignore-overlaps   <span class="built_in">disable</span> read-pair overlap detection</span><br><span class="line"></span><br><span class="line">Output options:</span><br><span class="line">  -o, --output FILE       write output to FILE [standard output]</span><br><span class="line">  -g, --BCF               generate genotype likelihoods <span class="keyword">in</span> BCF format</span><br><span class="line">  -v, --VCF               generate genotype likelihoods <span class="keyword">in</span> VCF format</span><br><span class="line"></span><br><span class="line">Output options <span class="keyword">for</span> mpileup format (without -g/-v):</span><br><span class="line">  -O, --output-BP         output base positions on reads</span><br><span class="line">  -s, --output-MQ         output mapping quality</span><br><span class="line"></span><br><span class="line">Output options <span class="keyword">for</span> genotype likelihoods (when -g/-v is used):</span><br><span class="line">  -t, --output-tags LIST  optional tags to output:</span><br><span class="line">               DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []</span><br><span class="line">  -u, --uncompressed      generate uncompressed VCF/BCF output</span><br><span class="line"></span><br><span class="line">SNP/INDEL genotype likelihoods options (effective with -g/-v):</span><br><span class="line">  -e, --ext-prob INT      Phred-scaled gap extension <span class="built_in">seq</span> error probability [20]</span><br><span class="line">  -F, --gap-frac FLOAT    minimum fraction of gapped reads [0.002]</span><br><span class="line">  -h, --tandem-qual INT   coefficient <span class="keyword">for</span> homopolymer errors [100]</span><br><span class="line">  -I, --skip-indels       <span class="keyword">do</span> not perform indel calling</span><br><span class="line">  -L, --max-idepth INT    maximum per-file depth <span class="keyword">for</span> INDEL calling [250]</span><br><span class="line">  -m, --min-ireads INT    minimum number gapped reads <span class="keyword">for</span> indel candidates [1]</span><br><span class="line">  -o, --open-prob INT     Phred-scaled gap open <span class="built_in">seq</span> error probability [40]</span><br><span class="line">  -p, --per-sample-mF     apply -m and -F per-sample <span class="keyword">for</span> increased sensitivity</span><br><span class="line">  -P, --platforms STR     comma separated list of platforms <span class="keyword">for</span> indels [all]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">Notes: Assuming diploid individuals.</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example: <code>samtools mpileup -C 50 -f ref.fasta -r chr3:1,000-2,000 in1.bam in2.bam</code></p>
</li>
<li><p>主要参数解读：</p>
<p>-A：在检测变异中，不忽略异常的reads对</p>
<p>-C：用于调节比对质量的系数，如果reads中含有过多的错配，不能设置为零</p>
<p>-D：输出每个样本的reads深度</p>
<p>-l：BED文件或者包含区域位点的位置列表文件</p>
<p>注意：位置文件包含两列，染色体和位置，从1开始计数。BED文件至少包含3列，染色体、起始和终止位置，开始端从0开始计数。</p>
<p>-r：在指定区域产生pileup，需已建立索引的bam文件，通常和-l参数一起使用</p>
<p>-o&#x2F;g&#x2F;v：输出文件类型（标准格式文件或者VCF、BCF文件）</p>
<p>-t：设置FORMAT和INFO的列表内容，以逗号分割</p>
<p>-u：生成未压缩的VCF和BCF文件</p>
<p>-I：跳过INDEL检测</p>
<p>-m：候选INDEL的最小间隔的reads</p>
<p>-f：输入有索引文件的fasta参考序列</p>
<p>-F ：含有间隔reads的最小片段</p>
</li>
<li><p>mpileup生成的结果包含6列：1) 参考序列名，2) 位置，3) 参考碱基，4) 比对上的reads数，5) 比对情况，6) 比对上的碱基的质量。其中第5列比较复杂,解释如下：<br> a) ‘.’代表与参考序列正链匹配。<br> b) ‘,’代表与参考序列负链匹配。<br> c) ‘ATCGN’代表在正链上的不匹配。<br> d) ‘atcgn’代表在负链上的不匹配。<br> e) ‘*’代表模糊碱基<br> f) ‘’代表匹配的碱基是一个read的开始；’’后面紧跟的ascii码减去33代表比对质量；这两个符号修饰的是后面的碱基，其后紧跟的碱基(.,ATCGatcgNn)代表该read的第一个碱基。<br> g) ‘$’代表一个read的结束，该符号修饰的是其前面的碱基。<br> h) 正则式’+[0-9]+[ACGTNacgtn]+’代表在该位点后插入的碱基；<br> i) 正则式’-[0-9]+[ACGTNacgtn]+’代表在该位点后缺失的碱基；</p>
</li>
</ul>
<h4 id="sort"><a href="#sort" class="headerlink" title="sort"></a>sort</h4><p>– sort alignment file 对比对后的bam文件进行排序，默认按coordinate进行排序</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools <span class="built_in">sort</span></span><br><span class="line"></span><br><span class="line">Usage: samtools <span class="built_in">sort</span> [options...] [in.bam]</span><br><span class="line">Options:</span><br><span class="line">  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)</span><br><span class="line">  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]</span><br><span class="line">  -n         Sort by <span class="built_in">read</span> name</span><br><span class="line">  -o FILE    Write final output to FILE rather than standard output</span><br><span class="line">  -T PREFIX  Write temporary files to PREFIX.nnnn.bam</span><br><span class="line">  -@, --threads INT</span><br><span class="line">             Set number of sorting and compression threads [1]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example: <code>samtools sort -T /tmp/aln.sorted -o aln.sorted.bam aln.bam</code></p>
</li>
<li><p>主要参数释义：</p>
<p>-l：设置文件压缩等级，0不压缩，9压缩最高</p>
<p>-m：每个线程运行内存大小（可使用K M G表示）</p>
<p>-n：按照read名称进行排序</p>
<p>-o：排序后的输出文件</p>
<p>-T：PREFIX临时文件前缀</p>
<p>-@：设置排序和压缩的线程数，默认单线程</p>
</li>
</ul>
<h4 id="split"><a href="#split" class="headerlink" title="split"></a>split</h4><ul>
<li><p>– splits a file by read group 根据read group标签将bam文件进行分割</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">$ samtools <span class="built_in">split</span></span><br><span class="line"></span><br><span class="line">Usage: samtools <span class="built_in">split</span> [-u &lt;unaccounted.bam&gt;[:&lt;unaccounted_header.sam&gt;]]</span><br><span class="line">                      [-f &lt;format_string&gt;] [-v] &lt;merged.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -f STRING       output filename format string [<span class="string">&quot;%*_%#.%.&quot;</span>]</span><br><span class="line">  -u FILE1        put reads with no RG tag or an unrecognised RG tag <span class="keyword">in</span> FILE1</span><br><span class="line">  -u FILE1:FILE2  ...and override the header with FILE2</span><br><span class="line">  -v              verbose output</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">Format string expansions:</span><br><span class="line">  %%     %</span><br><span class="line">  %*     <span class="built_in">basename</span></span><br><span class="line">  %<span class="comment">#     @RG index</span></span><br><span class="line">  %!     @RG ID</span><br><span class="line">  %.     filename extension <span class="keyword">for</span> output format</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools split merged.bam</code></li>
</ul>
</li>
</ul>
<h4 id="quickcheck"><a href="#quickcheck" class="headerlink" title="quickcheck"></a>quickcheck</h4><ul>
<li><p>– quickly check if SAM&#x2F;BAM&#x2F;CRAM file appears intact  检查SAM&#x2F;BAM&#x2F;CRAM文件的完整性</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools quickcheck</span><br><span class="line"></span><br><span class="line">Usage: samtools quickcheck [options] &lt;input&gt; [...]</span><br><span class="line">Options:</span><br><span class="line">  -v              verbose output (repeat <span class="keyword">for</span> more verbosity)</span><br><span class="line"></span><br><span class="line">Notes:</span><br><span class="line"></span><br><span class="line">1. In order to use this <span class="built_in">command</span> effectively, you should check its <span class="built_in">exit</span> status;</span><br><span class="line">   without any -v options it will NOT <span class="built_in">print</span> any output, even when some files</span><br><span class="line">   fail the check. One way to use quickcheck might be as a check that all</span><br><span class="line">   BAM files <span class="keyword">in</span> a directory are okay:</span><br><span class="line"></span><br><span class="line">        samtools quickcheck *.bam &amp;&amp; <span class="built_in">echo</span> <span class="string">&#x27;all ok&#x27;</span> \</span><br><span class="line">           || <span class="built_in">echo</span> <span class="string">&#x27;fail!&#x27;</span></span><br><span class="line"></span><br><span class="line">   To also determine <span class="built_in">which</span> files have failed, use the -v option:</span><br><span class="line"></span><br><span class="line">        samtools quickcheck -v *.bam &gt; bad_bams.fofn \</span><br><span class="line">           &amp;&amp; <span class="built_in">echo</span> <span class="string">&#x27;all ok&#x27;</span> \</span><br><span class="line">           || <span class="built_in">echo</span> <span class="string">&#x27;some files failed check, see bad_bams.fofn&#x27;</span></span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools quickcheck in1.bam in2.cram</code></li>
</ul>
</li>
</ul>
<h4 id="fastq"><a href="#fastq" class="headerlink" title="fastq"></a>fastq</h4><ul>
<li><p>– converts a BAM to a FASTQ 将bam格式文件转换为fastq文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">$ samtools fastq</span><br><span class="line"></span><br><span class="line">Usage: samtools fastq [options...] &lt;in.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -0 FILE   write paired reads flagged both or neither READ1 and READ2 to FILE</span><br><span class="line">  -1 FILE   write paired reads flagged READ1 to FILE</span><br><span class="line">  -2 FILE   write paired reads flagged READ2 to FILE</span><br><span class="line">  -f INT    only include reads with all bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -F INT    only include reads with none of the bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -n        don<span class="string">&#x27;t append /1 and /2 to the read name</span></span><br><span class="line"><span class="string">  -O        output quality in the OQ tag if present</span></span><br><span class="line"><span class="string">  -s FILE   write singleton reads to FILE [assume single-end]</span></span><br><span class="line"><span class="string">  -t        copy RG, BC and QT tags to the FASTQ header line</span></span><br><span class="line"><span class="string">  -v INT    default quality score if not given in file [1]</span></span><br><span class="line"><span class="string">      --input-fmt-option OPT[=VAL]</span></span><br><span class="line"><span class="string">               Specify a single input file format option in the form</span></span><br><span class="line"><span class="string">               of OPTION or OPTION=VALUE</span></span><br><span class="line"><span class="string">      --reference FILE</span></span><br><span class="line"><span class="string">               Reference sequence FASTA FILE [null]</span></span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools fastq input.bam &gt; output.fastq</code></li>
</ul>
</li>
</ul>
<h4 id="fasta"><a href="#fasta" class="headerlink" title="fasta"></a>fasta</h4><p> – converts a BAM to a FASTA 将bam格式文件转换为fasta文件</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">$ samtools fasta</span><br><span class="line"></span><br><span class="line">Usage: samtools fasta [options...] &lt;in.bam&gt;</span><br><span class="line">Options:</span><br><span class="line">  -0 FILE   write paired reads flagged both or neither READ1 and READ2 to FILE</span><br><span class="line">  -1 FILE   write paired reads flagged READ1 to FILE</span><br><span class="line">  -2 FILE   write paired reads flagged READ2 to FILE</span><br><span class="line">  -f INT    only include reads with all bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -F INT    only include reads with none of the bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -n        don<span class="string">&#x27;t append /1 and /2 to the read name</span></span><br><span class="line"><span class="string">  -s FILE   write singleton reads to FILE [assume single-end]</span></span><br><span class="line"><span class="string">  -t        copy RG, BC and QT tags to the FASTA header line</span></span><br><span class="line"><span class="string">      --input-fmt-option OPT[=VAL]</span></span><br><span class="line"><span class="string">               Specify a single input file format option in the form</span></span><br><span class="line"><span class="string">               of OPTION or OPTION=VALUE</span></span><br><span class="line"><span class="string">      --reference FILE</span></span><br><span class="line"><span class="string">               Reference sequence FASTA FILE [null]</span></span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools fasta input.bam &gt; output.fasta</code></li>
</ul>
<h3 id="2-4-数据统计-Statistics"><a href="#2-4-数据统计-Statistics" class="headerlink" title="2.4 数据统计(Statistics)"></a>2.4 数据统计(Statistics)</h3><h4 id="bedcov"><a href="#bedcov" class="headerlink" title="bedcov"></a>bedcov</h4><ul>
<li><p>– read depth per BED region 统计给定region区间内reads的深度，每个碱基的深度叠加在一起</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ samtools bedcov</span><br><span class="line"></span><br><span class="line">Usage: samtools bedcov [options] &lt;in.bed&gt; &lt;in1.bam&gt; [...]</span><br><span class="line"></span><br><span class="line">  -Q INT       Only count bases of at least INT quality [0]</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools bedcov in.bed aln.sorted.bam</code></li>
</ul>
</li>
</ul>
<h4 id="depth"><a href="#depth" class="headerlink" title="depth"></a>depth</h4><ul>
<li><p>– compute the depth 统计每个碱基位点的测序深度，注意计算前必须对bam文件排序并构建索引</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">$ samtools depth</span><br><span class="line"></span><br><span class="line">Usage: samtools depth [options] in1.bam [in2.bam [...]]</span><br><span class="line">Options:</span><br><span class="line">   -a                  output all positions (including zero depth)</span><br><span class="line">   -a -a (or -aa)      output absolutely all positions, including unused ref. sequences</span><br><span class="line">   -b &lt;bed&gt;            list of positions or regions</span><br><span class="line">   -f &lt;list&gt;           list of input BAM filenames, one per line [null]</span><br><span class="line">   -l &lt;int&gt;            <span class="built_in">read</span> length threshold (ignore reads shorter than &lt;int&gt;)</span><br><span class="line">   -d/-m &lt;int&gt;         maximum coverage depth [8000]</span><br><span class="line">   -q &lt;int&gt;            base quality threshold</span><br><span class="line">   -Q &lt;int&gt;            mapping quality threshold</span><br><span class="line">   -r &lt;chr:from-to&gt;    region</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br><span class="line"></span><br><span class="line">The output is a simple tab-separated table with three columns: reference name,</span><br><span class="line">position, and coverage depth.  Note that positions with zero coverage may be</span><br><span class="line">omitted by default; see the -a option.</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools depth aln.sorted.bam</code></li>
</ul>
</li>
</ul>
<h4 id="flagstat"><a href="#flagstat" class="headerlink" title="flagstat"></a>flagstat</h4><ul>
<li>– simple stats 统计bam文件中read的比对情况</li>
</ul>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flagstat</span><br><span class="line"></span><br><span class="line">Usage: samtools flagstat [--input-fmt-option OPT=VAL] &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools flagstat aln.sorted.bam</code></li>
</ul>
<h4 id="idxstats"><a href="#idxstats" class="headerlink" title="idxstats"></a>idxstats</h4><p> – BAM index stats  统计bam索引文件里的比对信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ samtools idxstats</span><br><span class="line"></span><br><span class="line">Usage: samtools idxstats &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools idxstats aln.sorted.bam</code></li>
<li>idxstats生成一个统计表格，共有4列，分别为”序列名，序列长度，比对上的reads数，unmapped reads number”。第4列应该是paired reads中有一端能匹配到该scaffold上，而另外一端不匹配到任何scaffolds上的reads数。</li>
</ul>
<h4 id="phase"><a href="#phase" class="headerlink" title="phase"></a>phase</h4><p>– Call and phase heterozygous SNPs</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">$ samtools phase</span><br><span class="line"></span><br><span class="line">Usage:   samtools phase [options] &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line">Options: -k INT    block length [13]</span><br><span class="line">         -b STR    prefix of BAMs to output [null]</span><br><span class="line">         -q INT    min het phred-LOD [37]</span><br><span class="line">         -Q INT    min base quality <span class="keyword">in</span> het calling [13]</span><br><span class="line">         -D INT    max <span class="built_in">read</span> depth [256]</span><br><span class="line">         -F        <span class="keyword">do</span> not attempt to fix chimeras</span><br><span class="line">         -A        drop reads with ambiguous phase</span><br><span class="line"></span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools phase test.bam</code></li>
</ul>
<h4 id="stats"><a href="#stats" class="headerlink" title="stats"></a>stats</h4><ul>
<li><p>– generate stats (former bamcheck)  对bam文件做详细统计, 统计结果可用mics&#x2F;plot-bamstats作图</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">$ samtools stats</span><br><span class="line"></span><br><span class="line">About: The program collects statistics from BAM files. The output can be visualized using plot-bamstats.</span><br><span class="line">Usage: samtools stats [OPTIONS] file.bam</span><br><span class="line">       samtools stats [OPTIONS] file.bam chr:from-to</span><br><span class="line">Options:</span><br><span class="line">    -c, --coverage &lt;int&gt;,&lt;int&gt;,&lt;int&gt;    Coverage distribution min,max,step [1,1000,1]</span><br><span class="line">    -d, --remove-dups                   Exclude from statistics reads marked as duplicates</span><br><span class="line">    -f, --required-flag  &lt;str|int&gt;      Required flag, 0 <span class="keyword">for</span> <span class="built_in">unset</span>. See also `samtools flags` [0]</span><br><span class="line">    -F, --filtering-flag &lt;str|int&gt;      Filtering flag, 0 <span class="keyword">for</span> <span class="built_in">unset</span>. See also `samtools flags` [0]</span><br><span class="line">        --GC-depth &lt;<span class="built_in">float</span>&gt;              the size of GC-depth bins (decreasing bin size increases memory requirement) [2e4]</span><br><span class="line">    -h, --<span class="built_in">help</span>                          This <span class="built_in">help</span> message</span><br><span class="line">    -i, --insert-size &lt;int&gt;             Maximum insert size [8000]</span><br><span class="line">    -I, --<span class="built_in">id</span> &lt;string&gt;                   Include only listed <span class="built_in">read</span> group or sample name</span><br><span class="line">    -l, --read-length &lt;int&gt;             Include <span class="keyword">in</span> the statistics only reads with the given <span class="built_in">read</span> length []</span><br><span class="line">    -m, --most-inserts &lt;<span class="built_in">float</span>&gt;          Report only the main part of inserts [0.99]</span><br><span class="line">    -P, --split-prefix &lt;str&gt;            Path or string prefix <span class="keyword">for</span> filepaths output by -S (default is input filename)</span><br><span class="line">    -q, --trim-quality &lt;int&gt;            The BWA trimming parameter [0]</span><br><span class="line">    -r, --ref-seq &lt;file&gt;                Reference sequence (required <span class="keyword">for</span> GC-depth and mismatches-per-cycle calculation).</span><br><span class="line">    -s, --sam                           Ignored (input format is auto-detected).</span><br><span class="line">    -S, --<span class="built_in">split</span> &lt;tag&gt;                   Also write statistics to separate files <span class="built_in">split</span> by tagged field.</span><br><span class="line">    -t, --target-regions &lt;file&gt;         Do stats <span class="keyword">in</span> these regions only. Tab-delimited file chr,from,to, 1-based, inclusive.</span><br><span class="line">    -x, --sparse                        Suppress outputting IS rows <span class="built_in">where</span> there are no insertions.</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools stats aln.sorted.bam</code></li>
<li>输出的信息比较多，部分如下：<br> Summary Numbers，raw total sequences，filtered sequences, reads mapped, reads mapped and paired,reads properly paired等信息<br> Fragment Qualitites：根据cycle统计每个位点上的碱基质量分布<br> Coverage distribution：深度为1，2，3，，，的碱基数目<br> ACGT content per cycle：ACGT在每个cycle中的比例<br> Insert sizes：插入长度的统计<br> Read lengths：read的长度分布</li>
</ul>
</li>
</ul>
<h3 id="2-5-可视化-Viewing"><a href="#2-5-可视化-Viewing" class="headerlink" title="2.5 可视化(Viewing)"></a>2.5 可视化(Viewing)</h3><h4 id="flags"><a href="#flags" class="headerlink" title="flags"></a>flags</h4><ul>
<li><p>– explain BAM flags   查看不同flag值的含义</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flags</span><br><span class="line"></span><br><span class="line">About: Convert between textual and numeric flag representation</span><br><span class="line">Usage: samtools flags INT|STR[,...]</span><br><span class="line"></span><br><span class="line">Flags:</span><br><span class="line">        0x1     PAIRED        .. paired-end (or multiple-segment) sequencing technology</span><br><span class="line">        0x2     PROPER_PAIR   .. each segment properly aligned according to the aligner</span><br><span class="line">        0x4     UNMAP         .. segment unmapped</span><br><span class="line">        0x8     MUNMAP        .. next segment <span class="keyword">in</span> the template unmapped</span><br><span class="line">        0x10    REVERSE       .. SEQ is reverse complemented</span><br><span class="line">        0x20    MREVERSE      .. SEQ of the next segment <span class="keyword">in</span> the template is reversed</span><br><span class="line">        0x40    READ1         .. the first segment <span class="keyword">in</span> the template</span><br><span class="line">        0x80    READ2         .. the last segment <span class="keyword">in</span> the template</span><br><span class="line">        0x100   SECONDARY     .. secondary alignment</span><br><span class="line">        0x200   QCFAIL        .. not passing quality controls</span><br><span class="line">        0x400   DUP           .. PCR or optical duplicate</span><br><span class="line">        0x800   SUPPLEMENTARY .. supplementary alignment</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example: <code>samtools flags PAIRED,UNMAP,MUNMAP</code></p>
</li>
<li><p><strong>FLAGS:</strong></p>
<table>
<thead>
<tr>
<th>0x1</th>
<th>PAIRED</th>
<th align="center">paired-end (or multiple-segment) sequencing technology</th>
</tr>
</thead>
<tbody><tr>
<td>0x2</td>
<td>PROPER_PAIR</td>
<td align="center">each segment properly aligned according to the aligner</td>
</tr>
<tr>
<td>0x4</td>
<td>UNMAP</td>
<td align="center">segment unmapped</td>
</tr>
<tr>
<td>0x8</td>
<td>MUNMAP</td>
<td align="center">next segment in the template unmapped</td>
</tr>
<tr>
<td>0x10</td>
<td>REVERSE</td>
<td align="center">SEQ is reverse complemented</td>
</tr>
<tr>
<td>0x20</td>
<td>MREVERSE</td>
<td align="center">SEQ of the next segment in the template is reverse complemented</td>
</tr>
<tr>
<td>0x40</td>
<td>READ1</td>
<td align="center">the first segment in the template</td>
</tr>
<tr>
<td>0x80</td>
<td>READ2</td>
<td align="center">the last segment in the template</td>
</tr>
<tr>
<td>0x100</td>
<td>SECONDARY</td>
<td align="center">secondary alignment</td>
</tr>
<tr>
<td>0x200</td>
<td>QCFAIL</td>
<td align="center">not passing quality controls</td>
</tr>
<tr>
<td>0x400</td>
<td>DUP</td>
<td align="center">PCR or optical duplicate</td>
</tr>
<tr>
<td>0x800</td>
<td>SUPPLEMENTARY</td>
<td align="center">supplementary alignment</td>
</tr>
</tbody></table>
</li>
</ul>
</li>
</ul>
<h4 id="tview"><a href="#tview" class="headerlink" title="tview"></a>tview</h4><p> – text alignment viewer  直观地显示reads比对到参考基因组的情况，类似于基因组浏览器</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">$ samtools tview</span><br><span class="line"></span><br><span class="line">Usage: samtools tview [options] &lt;aln.bam&gt; [ref.fasta]</span><br><span class="line">Options:</span><br><span class="line">   -d display      output as (H)tml or (C)urses or (T)ext</span><br><span class="line">   -p chr:pos      go directly to this position</span><br><span class="line">   -s STR          display only reads from this sample or group</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools tview aln.sorted.bam ref.fasta</code></li>
<li>需先对bam文件进行排序并构建索引</li>
</ul>
<h4 id="view"><a href="#view" class="headerlink" title="view"></a>view</h4><p>– SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion 将sam和bam文件进行格式互换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view</span><br><span class="line"></span><br><span class="line">Usage: samtools view [options] &lt;in.bam&gt;|&lt;in.sam&gt;|&lt;in.cram&gt; [region ...]</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -b       output BAM</span><br><span class="line">  -C       output CRAM (requires -T)</span><br><span class="line">  -1       use fast BAM compression (implies -b)</span><br><span class="line">  -u       uncompressed BAM output (implies -b)</span><br><span class="line">  -h       include header <span class="keyword">in</span> SAM output</span><br><span class="line">  -H       <span class="built_in">print</span> SAM header only (no alignments)</span><br><span class="line">  -c       <span class="built_in">print</span> only the count of matching records</span><br><span class="line">  -o FILE  output file name [stdout]</span><br><span class="line">  -U FILE  output reads not selected by filters to FILE [null]</span><br><span class="line">  -t FILE  FILE listing reference names and lengths (see long <span class="built_in">help</span>) [null]</span><br><span class="line">  -L FILE  only include reads overlapping this BED FILE [null]</span><br><span class="line">  -r STR   only include reads <span class="keyword">in</span> <span class="built_in">read</span> group STR [null]</span><br><span class="line">  -R FILE  only include reads with <span class="built_in">read</span> group listed <span class="keyword">in</span> FILE [null]</span><br><span class="line">  -q INT   only include reads with mapping quality &gt;= INT [0]</span><br><span class="line">  -l STR   only include reads <span class="keyword">in</span> library STR [null]</span><br><span class="line">  -m INT   only include reads with number of CIGAR operations consuming</span><br><span class="line">           query sequence &gt;= INT [0]</span><br><span class="line">  -f INT   only include reads with all bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -F INT   only include reads with none of the bits <span class="built_in">set</span> <span class="keyword">in</span> INT <span class="built_in">set</span> <span class="keyword">in</span> FLAG [0]</span><br><span class="line">  -x STR   <span class="built_in">read</span> tag to strip (repeatable) [null]</span><br><span class="line">  -B       collapse the backward CIGAR operation</span><br><span class="line">  -s FLOAT <span class="built_in">integer</span> part sets seed of random number generator [0];</span><br><span class="line">           rest sets fraction of templates to subsample [no subsampling]</span><br><span class="line">  -@, --threads INT</span><br><span class="line">           number of BAM/CRAM compression threads [0]</span><br><span class="line">  -?       <span class="built_in">print</span> long <span class="built_in">help</span>, including note about region specification</span><br><span class="line">  -S       ignored (input format is auto-detected)</span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -O, --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">  -T, --reference FILE</span><br><span class="line">               Reference sequence FASTA FILE [null]</span><br></pre></td></tr></table></figure>

<ul>
<li><p>Example ：<code>samtools view -bS test.sam &gt; test.bam</code></p>
</li>
<li><p>重要参数解读：</p>
<p>-b：输出bam格式</p>
<p>-C：输出CRAM文件</p>
<p>-1：快速压缩（需要-b）</p>
<p>-u：输出未压缩的bam文件，节约时间，占据较多磁盘空间（需要-b）</p>
<p>-h：默认输出sam文件不带表头，该参数设定后输出带表头信息</p>
<p>-H：仅仅输出表头信息</p>
<p>-c：仅打印匹配数</p>
<p>-o：输出文件（stdout标准输出）</p>
<p>-U：输出没有经过过滤选择的reads</p>
<p>-t：制表分隔符文件（需要提供额外的参考数据，比如参考基因组的索引）</p>
<p>-L：仅包括和bed文件有重叠的reads</p>
<p>-r：仅输出在STR读段组中的reads</p>
<p>-R：仅输出特定reads</p>
<p>-q：定位的质量大于INT[默认0]</p>
<p>-l：仅输出在STR 库中的reads</p>
<p>-F：获得比对上（mapped）的过滤设置[默认0]</p>
<p>-f：获得未必对上（unmapped）的过滤设置[默认0]</p>
<p>-T：使用fasta格式的参考序列</p>
</li>
</ul>
<h4 id="depad"><a href="#depad" class="headerlink" title="depad"></a>depad</h4><ul>
<li><p>– convert padded BAM to unpadded BAM</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools depad</span><br><span class="line"></span><br><span class="line">Usage:   samtools depad &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line">Options:</span><br><span class="line">  -s           Output is SAM (default is BAM)</span><br><span class="line">  -S           Input is SAM (default is BAM)</span><br><span class="line">  -u           Uncompressed BAM output (can<span class="string">&#x27;t use with -s)</span></span><br><span class="line"><span class="string">  -1           Fast compression BAM output (can&#x27;</span>t use with -s)</span><br><span class="line">  -T, --reference FILE</span><br><span class="line">               Padded reference sequence file [null]</span><br><span class="line">  -o FILE      Output file name [stdout]</span><br><span class="line">  -?           Longer <span class="built_in">help</span></span><br><span class="line">      --input-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single input file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br><span class="line">      --output-fmt FORMAT[,OPT[=VAL]]...</span><br><span class="line">               Specify output format (SAM, BAM, CRAM)</span><br><span class="line">      --output-fmt-option OPT[=VAL]</span><br><span class="line">               Specify a single output file format option <span class="keyword">in</span> the form</span><br><span class="line">               of OPTION or OPTION=VALUE</span><br></pre></td></tr></table></figure>

<ul>
<li>Example: <code>samtools depad input.bam</code></li>
</ul>
</li>
</ul>
<h3 id="2-6-常用示例"><a href="#2-6-常用示例" class="headerlink" title="2.6 常用示例"></a>2.6 常用示例</h3><p>sam&#x2F;bam格式文件互换</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bS -1 test.sam &gt; test.bam <span class="comment"># sam转bam</span></span><br><span class="line">$ samtools view -h test.bam &gt; test.sam <span class="comment"># bam转sam</span></span><br></pre></td></tr></table></figure>

<p>提取比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bF 4 test.bam &gt; test.F.bam</span><br></pre></td></tr></table></figure>

<p>提取没有比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bf 4 test.bam &gt; test.f.bam</span><br></pre></td></tr></table></figure>

<p>双端reads都比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bF 12 test.bam &gt; test.12.bam</span><br></pre></td></tr></table></figure>

<p>单端reads1比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools view -bF 4 -f 8  <span class="built_in">test</span> .bam &gt; test1.bam</span><br></pre></td></tr></table></figure>

<p>单端reads2比对到参考基因组上的数据</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -bF 8 -f 4 test.bam &gt; test2.bam</span><br></pre></td></tr></table></figure>

<p>提取bam文件中比对到scaffold1上的比对结果，并保存到sam文件格式</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view abc.bam scaffold1 &gt; scaffold1.sam</span><br></pre></td></tr></table></figure>

<p>提取scaffold1上能比对到30k到100k区域的比对结果</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view abc.bam scaffold1:30000-100000 &gt; scaffold1_30k-100k.sam</span><br></pre></td></tr></table></figure>

<p>根据fasta文件，将 header 加入到 sam 或 bam 文件中</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools view -T genome.fasta -h scaffold1.sam &gt; scaffold1.h.sam</span><br></pre></td></tr></table></figure>

<p>call SNP和INDEL等变异信息</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ samtools mpileup -f genome.fasta abc.bam &gt; abc.txt</span><br><span class="line">$ samtools mpileup -gSDf genome.fasta abc.bam &gt; abc.bcf</span><br><span class="line">$ samtools mpileup -guSDf genome.fasta abc.bam | \</span><br><span class="line">           bcftools view -cvNg - &gt; abc.vcf</span><br></pre></td></tr></table></figure>
    </div>

    
    
    
        <div class="reward-container">
  <div></div>
  <button onclick="var qr = document.getElementById('qr'); qr.style.display = (qr.style.display === 'none') ? 'block' : 'none';">
    打赏
  </button>
  <div id="qr" style="display: none;">

  </div>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/%E5%B7%A5%E5%85%B7/" rel="tag"># 工具</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2023/02/27/%E8%82%BF%E7%98%A4%E6%A3%80%E6%B5%8B-%E5%9F%BA%E7%A1%80%E6%A6%82%E5%BF%B5/" rel="prev" title="肿瘤检测-基础概念">
      <i class="fa fa-chevron-left"></i> 肿瘤检测-基础概念
    </a></div>
      <div class="post-nav-item"></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#samtools-%E5%B8%B8%E7%94%A8%E5%91%BD%E4%BB%A4"><span class="nav-number">1.</span> <span class="nav-text">samtools 常用命令</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-%E8%BD%AF%E4%BB%B6%E4%BB%8B%E7%BB%8D%E5%8F%8A%E5%AE%89%E8%A3%85"><span class="nav-number">1.1.</span> <span class="nav-text">1.软件介绍及安装</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-Samtools-%E5%9F%BA%E6%9C%AC%E5%91%BD%E4%BB%A4"><span class="nav-number">1.2.</span> <span class="nav-text">2.Samtools 基本命令</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#2-1-%E6%9E%84%E5%BB%BA%E7%B4%A2%E5%BC%95"><span class="nav-number">1.2.1.</span> <span class="nav-text">2.1 构建索引</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#2-1-1-dict"><span class="nav-number">1.2.1.1.</span> <span class="nav-text">2.1.1 dict</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-1-2-faidx"><span class="nav-number">1.2.1.2.</span> <span class="nav-text">2.1.2  faidx</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-1-3-index"><span class="nav-number">1.2.1.3.</span> <span class="nav-text">2.1.3 index</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-2-%E6%96%87%E4%BB%B6%E7%BC%96%E8%BE%91-Editing"><span class="nav-number">1.2.2.</span> <span class="nav-text">2.2 文件编辑(Editing)</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#calmd"><span class="nav-number">1.2.2.1.</span> <span class="nav-text">calmd</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#fixmate"><span class="nav-number">1.2.2.2.</span> <span class="nav-text">fixmate</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#reheader"><span class="nav-number">1.2.2.3.</span> <span class="nav-text">reheader</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#rmdup"><span class="nav-number">1.2.2.4.</span> <span class="nav-text">rmdup</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#targetcut"><span class="nav-number">1.2.2.5.</span> <span class="nav-text">targetcut</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#addreplacerg"><span class="nav-number">1.2.2.6.</span> <span class="nav-text">addreplacerg</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-3-%E6%96%87%E4%BB%B6%E5%A4%84%E7%90%86-File-operations"><span class="nav-number">1.2.3.</span> <span class="nav-text">2.3 文件处理(File operations)</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#collate"><span class="nav-number">1.2.3.1.</span> <span class="nav-text">collate</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#cat"><span class="nav-number">1.2.3.2.</span> <span class="nav-text">cat</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#merge"><span class="nav-number">1.2.3.3.</span> <span class="nav-text">merge</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#mpileup"><span class="nav-number">1.2.3.4.</span> <span class="nav-text">mpileup</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#sort"><span class="nav-number">1.2.3.5.</span> <span class="nav-text">sort</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#split"><span class="nav-number">1.2.3.6.</span> <span class="nav-text">split</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#quickcheck"><span class="nav-number">1.2.3.7.</span> <span class="nav-text">quickcheck</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#fastq"><span class="nav-number">1.2.3.8.</span> <span class="nav-text">fastq</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#fasta"><span class="nav-number">1.2.3.9.</span> <span class="nav-text">fasta</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-4-%E6%95%B0%E6%8D%AE%E7%BB%9F%E8%AE%A1-Statistics"><span class="nav-number">1.2.4.</span> <span class="nav-text">2.4 数据统计(Statistics)</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#bedcov"><span class="nav-number">1.2.4.1.</span> <span class="nav-text">bedcov</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#depth"><span class="nav-number">1.2.4.2.</span> <span class="nav-text">depth</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#flagstat"><span class="nav-number">1.2.4.3.</span> <span class="nav-text">flagstat</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#idxstats"><span class="nav-number">1.2.4.4.</span> <span class="nav-text">idxstats</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#phase"><span class="nav-number">1.2.4.5.</span> <span class="nav-text">phase</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#stats"><span class="nav-number">1.2.4.6.</span> <span class="nav-text">stats</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-5-%E5%8F%AF%E8%A7%86%E5%8C%96-Viewing"><span class="nav-number">1.2.5.</span> <span class="nav-text">2.5 可视化(Viewing)</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#flags"><span class="nav-number">1.2.5.1.</span> <span class="nav-text">flags</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#tview"><span class="nav-number">1.2.5.2.</span> <span class="nav-text">tview</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#view"><span class="nav-number">1.2.5.3.</span> <span class="nav-text">view</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#depad"><span class="nav-number">1.2.5.4.</span> <span class="nav-text">depad</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-6-%E5%B8%B8%E7%94%A8%E7%A4%BA%E4%BE%8B"><span class="nav-number">1.2.6.</span> <span class="nav-text">2.6 常用示例</span></a></li></ol></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Chao Zhang"
      src="/images/avatar.png">
  <p class="site-author-name" itemprop="name">Chao Zhang</p>
  <div class="site-description" itemprop="description">python, shell, R, perl, java</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">4</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">6</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">标签</span>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/chaopower-zhang" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;chaopower-zhang" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="/chaopower@126.com" title="E-Mail → chaopower@126.com"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>
  <div class="cc-license motion-element" itemprop="license">
    <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" class="cc-opacity" rel="noopener" target="_blank"><img src="/images/cc-by-nc-sa.svg" alt="Creative Commons"></a>
  </div>



      </div>
        <div class="back-to-top motion-element">
          <i class="fa fa-arrow-up"></i>
          <span>0%</span>
        </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2023</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">chaopower@126.com</span>
</div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

</body>
</html>
